// office-viewer/lib/excel/render/autoFilter/filterColumn/FiltersUI.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var H = require('../../../../util/H.js');
var CheckBoxList = require('../../ui/CheckBoxList.js');
var getFilterValues = require('./getFilterValues.js');
var setFilterValues = require('./setFilterValues.js');

/**
 * 按文本分类进行过滤
 */
var FiltersUI = /** @class */ (function () {
    function FiltersUI(autoFilterIcon, container, sheet, autoFilter, colIndex, headerRowCount, texts) {
        var filtersContainer = H.H('div', {
            className: 'ov-excel-auto-filter__menu-filter-filters',
            parent: container
        });
        this.filtersContainer = filtersContainer;
        var filterValues = getFilterValues.getFilterValues(autoFilter, colIndex);
        var selectAll = filterValues.size === 0;
        var options = texts.map(function (text) {
            return {
                text: text,
                value: text,
                checked: selectAll || filterValues.has(text)
            };
        });
        var selectAllText = sheet.translate('checkBox.selectAll');
        var searchPlaceholder = sheet.translate('checkBox.search');
        new CheckBoxList.CheckBoxList(filtersContainer, selectAllText, searchPlaceholder, options, function (checked, option, options) {
            setFilterValues.setFilterValues(autoFilter, colIndex, options);
            sheet.applyAutoFilter(autoFilter, headerRowCount);
            sheet.getWorkbook().uiEvent.emit('APPLY_AUTO_FILTER', sheet.getIndex());
            autoFilterIcon.syncFilterIcon();
        });
    }
    return FiltersUI;
}());

exports.FiltersUI = FiltersUI;
