// office-viewer/lib/excel/render/cell/frozen/getFrozenTopViewPointRange.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var getViewRange = require('../../../sheet/getViewRange.js');
var getFrozenRange = require('./getFrozenRange.js');

/**
 * 获取顶部冻结的范围
 * @param ySplit
 * @param width
 * @param scrollLeft
 * @param getRowHeight
 * @param getColWidth
 * @param colPositionCache
 * @param colHiddenRange
 * @returns
 */
function getFrozenTopViewPointRange(ySplit, width, scrollLeft, leftShift, topShift, getRowHeight, getColWidth, colPositionCache, colHiddenRange) {
    if (colPositionCache === void 0) { colPositionCache = []; }
    if (colHiddenRange === void 0) { colHiddenRange = []; }
    var _a = getViewRange.getRange(scrollLeft, leftShift, width, getColWidth, colHiddenRange, colPositionCache), cols = _a.indexes, startColOffset = _a.startOffset, colSizes = _a.sizes;
    var _b = getFrozenRange.getFrozenRange(ySplit, topShift, getRowHeight), rows = _b.indexes, startRowOffset = _b.startOffset, rowSizes = _b.sizes, height = _b.length;
    return {
        region: 'top-frozen',
        rows: rows,
        rowSizes: rowSizes,
        height: height,
        startRowOffset: startRowOffset,
        cols: cols,
        colSizes: colSizes,
        width: width,
        startColOffset: startColOffset
    };
}

exports.getFrozenTopViewPointRange = getFrozenTopViewPointRange;
