// office-viewer/esm/excel/render/selection/hitTest.js
import { hitTestInRange } from './hitTestInRange.js';

/**
 * 判断当前点击位置下是什么
 * @param offsetX 鼠标点击的 x 坐标
 * @param offsetY 鼠标点击的 y 坐标
 * @param rowHeaderWidth 表头行高
 * @param colHeaderHeight 表头列宽
 * @param viewRange 内容显示区的范围
 * @param frozenViewRange 冻结区域的范围
 */
function hitTest(offsetX, offsetY, rowHeaderWidth, colHeaderHeight, gridLineHitRange, viewRange, frozenViewRange, mergeCells) {
    // 点击到表头左上角，这个行为是选中所有行和列
    if (offsetX < rowHeaderWidth && offsetY < colHeaderHeight) {
        return {
            type: 'corner',
            region: 'normal',
            startRow: 0,
            startCol: 0,
            endCol: 0,
            endRow: 0,
            x: 0,
            y: 0,
            width: rowHeaderWidth,
            height: colHeaderHeight
        };
    }
    if (frozenViewRange) {
        var topViewRange = frozenViewRange.topViewRange, leftViewRange = frozenViewRange.leftViewRange, topLeftViewRange = frozenViewRange.topLeftViewRange;
        var hitTestTopLeft = hitTestInRange('top-left-frozen', topLeftViewRange, offsetX, offsetY, rowHeaderWidth, colHeaderHeight, gridLineHitRange, mergeCells);
        if (hitTestTopLeft) {
            return hitTestTopLeft;
        }
        var hitTestLeft = hitTestInRange('left-frozen', leftViewRange, offsetX, offsetY, rowHeaderWidth, colHeaderHeight, gridLineHitRange, mergeCells);
        if (hitTestLeft) {
            return hitTestLeft;
        }
        var hitTestTop = hitTestInRange('top-frozen', topViewRange, offsetX, offsetY, rowHeaderWidth, colHeaderHeight, gridLineHitRange, mergeCells);
        if (hitTestTop) {
            return hitTestTop;
        }
    }
    return hitTestInRange('normal', viewRange, offsetX, offsetY, rowHeaderWidth, colHeaderHeight, gridLineHitRange, mergeCells);
}

export { hitTest };
