// office-viewer/lib/excel/render/ui/CheckBox.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var H = require('../../../util/H.js');

var CheckBox = /** @class */ (function () {
    function CheckBox(container, option, onChange) {
        this.option = option;
        var checkboxContainer = H.H('div', {
            className: 'excel-checkbox-container',
            parent: container
        });
        var label = H.H('label', {
            parent: checkboxContainer
        });
        this.checkboxContainer = checkboxContainer;
        var checkbox = H.H('input', {
            type: 'checkbox',
            parent: label
        });
        this.checkbox = checkbox;
        checkbox.checked = option.checked;
        checkbox.onchange = function () {
            option.checked = checkbox.checked;
            onChange(checkbox.checked, option);
        };
        H.H('span', {
            innerText: option.text,
            parent: label
        });
    }
    CheckBox.prototype.getText = function () {
        return this.option.text;
    };
    CheckBox.prototype.setChecked = function (checked) {
        this.option.checked = checked;
        this.checkbox.checked = checked;
    };
    CheckBox.prototype.isChecked = function () {
        return this.checkbox.checked;
    };
    CheckBox.prototype.show = function () {
        this.checkboxContainer.style.display = 'block';
    };
    CheckBox.prototype.hide = function () {
        this.checkboxContainer.style.display = 'none';
    };
    return CheckBox;
}());

exports.CheckBox = CheckBox;
