// office-viewer/lib/word/parse/parseTablePr.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var OpenXML = require('../../OpenXML.js');
var parseBorder = require('./parseBorder.js');
var parseInsideBorders = require('./parseInsideBorders.js');
var parseTblWidth = require('./parseTblWidth.js');
var parseColor = require('./parseColor.js');
var parseSize = require('./parseSize.js');
var parseTcPr = require('./parseTcPr.js');
var parseCellMargin = require('./parseCellMargin.js');

/**
 * 表格的 jc 需要使用 float 来实现
 * http://officeopenxml.com/WPtableAlignment.php
 */
function parseTblJc(element, cssStyle) {
    var val = OpenXML.getVal(element);
    switch (val) {
        case 'left':
        case 'start':
            // TODO: 会导致前面的文字掉下去，感觉还是不能支持这个功能
            // cssStyle['float'] = 'left';
            break;
        case 'right':
        case 'end':
            cssStyle['float'] = 'right';
    }
}
/**
 * 这个其实分左右，但目前只支持左，右可能是阿拉伯语？
 * http://webapp.docx4java.org/OnlineDemo/ecma376/WordML/tblInd_2.html
 */
function parseTblInd(element, style) {
    var width = parseTblWidth.parseTblWidth(element);
    if (width) {
        style['margin-left'] = width;
    }
}
function parseTblW(element, style) {
    var width = parseTblWidth.parseTblWidth(element);
    if (width) {
        style['width'] = width;
    }
}
// http://webapp.docx4java.org/OnlineDemo/ecma376/WordML/ST_TblStyleOverrideType.html
// val 是旧的格式
function parseTblLook(child) {
    var tblLook = {};
    var tblLookVal = OpenXML.getValHex(child);
    if (OpenXML.getAttrBoolean(child, 'firstRow', false) || tblLookVal & 0x0020) {
        tblLook['firstRow'] = true;
    }
    if (OpenXML.getAttrBoolean(child, 'lastRow', false) || tblLookVal & 0x0040) {
        tblLook['lastRow'] = true;
    }
    if (OpenXML.getAttrBoolean(child, 'firstColumn', false) || tblLookVal & 0x0080) {
        tblLook['firstColumn'] = true;
    }
    if (OpenXML.getAttrBoolean(child, 'lastColumn', false) || tblLookVal & 0x0100) {
        tblLook['lastColumn'] = true;
    }
    if (OpenXML.getAttrBoolean(child, 'noHBand', false) || tblLookVal & 0x0200) {
        tblLook['noHBand'] = true;
    }
    else {
        tblLook['noHBand'] = false;
    }
    if (OpenXML.getAttrBoolean(child, 'noVBand', false) || tblLookVal & 0x0400) {
        tblLook['noVBand'] = true;
    }
    else {
        tblLook['noVBand'] = false;
    }
    return tblLook;
}
/**
 * http://webapp.docx4java.org/OnlineDemo/ecma376/WordML/tblpPr.html
 * 只支持部分
 */
function parseTblpPr(word, child, style) {
    // 如果设置 padding 会导致绝对定位不准确，所以一旦设置就不支持
    if (typeof word.renderOptions.padding === 'undefined') {
        var tplpX = parseSize.parseSize(child, 'w:tblpX');
        var tplpY = parseSize.parseSize(child, 'w:tblpY');
        // style.position = 'absolute';
        style.top = tplpY;
        style.left = tplpX;
    }
    // 之前想用 float 来实现，但是会导致文字掉下去
    // const topFromText = parseSize(child, 'w:topFromText');
    // const bottomFromText = parseSize(child, 'w:bottomFromText');
    // const rightFromText = parseSize(child, 'w:rightFromText');
    // const leftFromText = parseSize(child, 'w:leftFromText');
    // style['float'] = 'left';
    // style['margin-bottom'] = addSize(style['margin-bottom'], bottomFromText);
    // style['margin-left'] = addSize(style['margin-left'], leftFromText);
    // style['margin-right'] = addSize(style['margin-right'], rightFromText);
    // style['margin-top'] = addSize(style['margin-top'], topFromText);
}
/**
 * http://officeopenxml.com/WPtableLayout.php
 */
function parseTblLayout(element, style) {
    var type = element.getAttribute('w:type');
    if (type === 'fixed') {
        style['table-layout'] = 'fixed';
    }
}
function parseTablePr(word, element) {
    var e_1, _a;
    var properties = {};
    var tableStyle = {};
    var tcStyle = {};
    properties.tblLook = {};
    properties.cssStyle = tableStyle;
    properties.tcCSSStyle = tcStyle;
    try {
        for (var _b = tslib.__values(element.children), _c = _b.next(); !_c.done; _c = _b.next()) {
            var child = _c.value;
            var tagName = child.tagName;
            switch (tagName) {
                case 'w:tblBorders':
                    parseBorder.parseBorders(word, child, tableStyle);
                    properties.insideBorder = parseInsideBorders.parseInsideBorders(word, child);
                    break;
                case 'w:tcBorders':
                    parseBorder.parseBorders(word, child, tableStyle);
                    break;
                case 'w:tblInd':
                    parseTblInd(child, tableStyle);
                    break;
                case 'w:jc':
                    parseTblJc(child, tableStyle);
                    break;
                case 'w:tblCellMar':
                case 'w:tcMar':
                    // http://webapp.docx4java.org/OnlineDemo/ecma376/WordML/tblCellMar_1.html
                    parseCellMargin.parseCellMargin(child, tcStyle);
                    break;
                case 'w:tblStyle':
                    properties.pStyle = OpenXML.getVal(child);
                    break;
                case 'w:tblW':
                    parseTblW(child, tableStyle);
                    break;
                case 'w:shd':
                    // http://officeopenxml.com/WPtableShading.php
                    tableStyle['background-color'] = parseColor.parseShdColor(word, child);
                    break;
                case 'w:tblCaption':
                    properties.tblCaption = OpenXML.getVal(child);
                    break;
                case 'w:tblCellSpacing':
                    parseTcPr.parseTblCellSpacing(child, tableStyle);
                    break;
                case 'w:tblLayout':
                    parseTblLayout(child, tableStyle);
                    break;
                case 'w:tblLook':
                    properties.tblLook = parseTblLook(child);
                    break;
                case 'w:tblStyleRowBandSize':
                    properties.rowBandSize = OpenXML.getValNumber(child);
                    break;
                case 'w:tblStyleColBandSize':
                    properties.colBandSize = OpenXML.getValNumber(child);
                    break;
                case 'w:tblpPr':
                    parseTblpPr(word, child, tableStyle);
                    break;
                default:
                    console.warn('parseTableProperties unknown tag', tagName, child);
            }
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
        }
        finally { if (e_1) throw e_1.error; }
    }
    return properties;
}

exports.parseTablePr = parseTablePr;
