// office-viewer/lib/word/render/renderStyle.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var dom = require('../../util/dom.js');

/**
 * 渲染内置样式及自定义样式
 */
/**
 * 文档基础默认样式
 */
function generateDefaultStyle(word) {
    var _a;
    var styles = word.styles;
    var defaultStyle = styles.defaultStyle;
    var defaultPStyle = '';
    if (defaultStyle === null || defaultStyle === void 0 ? void 0 : defaultStyle.pPr) {
        defaultPStyle = dom.styleToText(defaultStyle.pPr.cssStyle);
    }
    var defaultRStyle = '';
    if (defaultStyle === null || defaultStyle === void 0 ? void 0 : defaultStyle.rPr) {
        defaultRStyle = dom.styleToText(defaultStyle.rPr.cssStyle);
    }
    var hyphens = ((_a = word.settings) === null || _a === void 0 ? void 0 : _a.autoHyphenation) ? 'hyphens: auto;' : '';
    var classPrefix = word.getClassPrefix();
    return "\n\n\n  /** docDefaults **/\n  .".concat(classPrefix, " {\n    --docx-theme-font-minorHAnsi: Calibri,  Helvetica, Arial, 'Helvetica Neue';\n    --docx-theme-font-minorEastAsia: 'PingFang SC', 'Microsoft YaHei', 'Hiragino Sans GB', 'STHeiti',\n    'Microsoft YaHei';\n  }\n\n  .").concat(classPrefix, " p {\n    margin: 0;\n    padding: 0;\n    line-height: 1.5;\n    ").concat(hyphens, "\n  }\n\n  .").concat(classPrefix, " .justify {\n    text-align-last: left;\n  }\n\n  .").concat(classPrefix, " table {\n    border-spacing: 0;\n  }\n\n  .").concat(classPrefix, " .").concat(classPrefix, "-p {\n    ").concat(defaultPStyle, "\n  }\n\n  .").concat(classPrefix, " .").concat(classPrefix, "-r {\n    overflow-wrap: break-word;\n    ").concat(defaultRStyle, "\n  }\n  ");
}
/**
 * 生成表格级别样式
 */
function generateTableStyle(classPrefix, styleDisplayId, style) {
    var tblStyleText = '';
    var tblPr = style.tblPr;
    var tcPr = style.tcPr;
    if (tblPr) {
        var tblStyle = dom.styleToText(tblPr.cssStyle);
        var tblTcStyle = dom.styleToText(tblPr.tcCSSStyle);
        tblStyleText += "\n .".concat(classPrefix, " .").concat(styleDisplayId, " {\n  border-collapse: collapse;\n  ").concat(tblStyle, "\n }\n\n .").concat(classPrefix, " .").concat(styleDisplayId, " > tbody > tr > td {\n  ").concat(tblTcStyle, "\n }\n ");
        if (tblPr.insideBorder) {
            var insideBorder = tblPr.insideBorder;
            if (insideBorder.H) {
                tblStyleText += "\n      .".concat(classPrefix, " .").concat(styleDisplayId, " > tbody > tr > td {\n        border-top: ").concat(insideBorder.H, ";\n      }");
            }
            if (insideBorder.V) {
                tblStyleText += "\n      .".concat(classPrefix, " .").concat(styleDisplayId, " > tbody > tr > td {\n        border-left: ").concat(insideBorder.V, ";\n      }");
            }
        }
    }
    if (tcPr) {
        var tcStyle = dom.styleToText(tcPr.cssStyle);
        tblStyleText += "\n    .".concat(classPrefix, " .").concat(styleDisplayId, " > tbody > tr > td {\n     ").concat(tcStyle, "\n    }\n    ");
    }
    return tblStyleText;
}
// 用于生成表格 override 相关的样式，用于行或列
function genTblOverrideStyle(tblPrefix, classPrefix, overrideType, tblStylePrStyle) {
    var _a, _b, _c, _d, _e, _f;
    var styleText = '';
    var trStyle = dom.styleToText((_a = tblStylePrStyle.trPr) === null || _a === void 0 ? void 0 : _a.cssStyle);
    var enableType = '';
    // 在 tblLook 里可以通过这些属性来控制是否启用
    switch (overrideType) {
        case 'firstCol':
            enableType = 'enable-firstColumn';
            break;
        case 'lastCol':
            enableType = 'enable-lastColumn';
            break;
        case 'firstRow':
            enableType = 'enable-firstRow';
            break;
        case 'lastRow':
            enableType = 'enable-lastRow';
            break;
        case 'band1Horz':
        case 'band2Horz':
            enableType = 'enable-hBand';
            break;
        case 'band1Vert':
        case 'band2Vert':
            enableType = 'enable-vBand';
            break;
    }
    if (trStyle) {
        styleText += "\n    ".concat(tblPrefix, ".").concat(enableType, " > tbody > tr.").concat(overrideType, "{\n       ").concat(trStyle, "\n    }\n    ");
    }
    var tcStyle = dom.styleToText((_b = tblStylePrStyle.tcPr) === null || _b === void 0 ? void 0 : _b.cssStyle);
    if (tcStyle) {
        styleText += "\n    ".concat(tblPrefix, ".").concat(enableType, " > tbody > tr > td.").concat(overrideType, " {\n       ").concat(tcStyle, "\n    }\n    ");
        if ((_c = tblStylePrStyle.tcPr) === null || _c === void 0 ? void 0 : _c.insideBorder) {
            var insideBorder = (_d = tblStylePrStyle.tcPr) === null || _d === void 0 ? void 0 : _d.insideBorder;
            if (insideBorder.H) {
                styleText += "\n          ".concat(tblPrefix, ".").concat(enableType, " > tbody > tr > td.").concat(overrideType, " {\n            border-top: ").concat(insideBorder.H, ";\n          }");
            }
            if (insideBorder.V) {
                // 这个主要是为了应对 GridTable5Dark-Accent5 里 firstRow 的情况，它其实有 right 设置，也得去掉
                if (insideBorder.V === 'none') {
                    styleText += "\n          ".concat(tblPrefix, ".").concat(enableType, " > tbody > tr > td.").concat(overrideType, " {\n            border-left: none;\n            border-right: none;\n          }");
                }
                else {
                    styleText += "\n          ".concat(tblPrefix, ".").concat(enableType, " > tbody > tr > td.").concat(overrideType, " {\n            border-left: ").concat(insideBorder.V, ";\n          }");
                }
            }
        }
    }
    var pStyle = dom.styleToText((_e = tblStylePrStyle.pPr) === null || _e === void 0 ? void 0 : _e.cssStyle);
    if (pStyle) {
        styleText += "\n    ".concat(tblPrefix, ".").concat(enableType, " > tbody > tr > td.").concat(overrideType, " > .").concat(classPrefix, "-p {\n       ").concat(pStyle, "\n    }\n    ");
    }
    var rStyle = dom.styleToText((_f = tblStylePrStyle.rPr) === null || _f === void 0 ? void 0 : _f.cssStyle);
    if (rStyle) {
        styleText += "\n    ".concat(tblPrefix, ".").concat(enableType, " > tbody > tr > td.").concat(overrideType, " > .").concat(classPrefix, "-p > .").concat(classPrefix, "-r {\n       ").concat(rStyle, "\n    }\n    ");
    }
    return styleText;
}
// 表格覆盖样式的顺序，权重高的放后面
var overrideTypeOrder = new Set([
    'wholeTable',
    'band1Horz',
    'band2Horz',
    'band1Vert',
    'band2Vert',
    'firstCol',
    'firstRow',
    'lastCol',
    'lastRow',
    'neCell',
    'nwCell',
    'seCell',
    'swCell'
]);
// 生成表格覆盖样式
function genOverrideTblStylePr(classPrefix, styleDisplayId, tblStylePr) {
    var e_1, _a;
    if (!tblStylePr) {
        return '';
    }
    var tblStylePrText = '';
    var stylePrefix = ".".concat(classPrefix, " .").concat(styleDisplayId);
    try {
        for (var overrideTypeOrder_1 = tslib.__values(overrideTypeOrder), overrideTypeOrder_1_1 = overrideTypeOrder_1.next(); !overrideTypeOrder_1_1.done; overrideTypeOrder_1_1 = overrideTypeOrder_1.next()) {
            var overrideType = overrideTypeOrder_1_1.value;
            if (overrideType in tblStylePr) {
                var overrideStylePr = tblStylePr[overrideType];
                tblStylePrText += genTblOverrideStyle(stylePrefix, classPrefix, overrideType, overrideStylePr);
            }
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (overrideTypeOrder_1_1 && !overrideTypeOrder_1_1.done && (_a = overrideTypeOrder_1.return)) _a.call(overrideTypeOrder_1);
        }
        finally { if (e_1) throw e_1.error; }
    }
    return tblStylePrText;
}
/**
 * 生成样式类
 */
function generateStyle(word) {
    var styles = word.styles;
    var styleMap = styles.styleMap;
    var classPrefix = word.getClassPrefix();
    var styleResult = '';
    for (var styleId in styleMap) {
        var styleDisplayId = word.getStyleIdDisplayName(styleId);
        var styleData = styleMap[styleId];
        var pPr = styleData.pPr;
        var pStyleText = '';
        if (pPr) {
            var pStyle = dom.styleToText(pPr.cssStyle);
            pStyleText = "\n      .".concat(classPrefix, " .").concat(styleDisplayId, " {\n        ").concat(pStyle, "\n      }\n      ");
        }
        var rStyleText = '';
        if (styleData.rPr) {
            var rStyle = dom.styleToText(styleData.rPr.cssStyle);
            rStyleText = "\n      .".concat(classPrefix, " .").concat(styleDisplayId, " > .").concat(classPrefix, "-r {\n        ").concat(rStyle, "\n      }\n      ");
        }
        var tblStyleText = generateTableStyle(classPrefix, styleDisplayId, styleData);
        var tblStylePr = genOverrideTblStylePr(classPrefix, styleDisplayId, styleData.tblStylePr);
        styleResult += "\n    ".concat(pStyleText, "\n    ").concat(rStyleText, "\n    ").concat(tblStyleText, "\n    ").concat(tblStylePr, "\n    ");
    }
    return styleResult;
}
/**
 * 渲染所有样式
 */
function renderStyle(word) {
    var style = dom.createElement('style');
    var docDefaults = generateDefaultStyle(word);
    var styleText = generateStyle(word);
    style.textContent = "\n  ".concat(docDefaults, "\n\n  ").concat(styleText, "\n  ");
    return style;
}

exports.generateTableStyle = generateTableStyle;
exports.renderStyle = renderStyle;
