// qrcode-react-next/es/QRCodeSvg.js
import t from "react";
import v from "./hooks/useQrCode.js";
function $(h) {
  const { value: m, config: s, styleConfig: n, logoConfig: f, className: a } = h, {
    size: r = 200,
    margin: w = 0,
    color: o = "#000",
    bgColor: i = "#fff",
    eyeBorderColor: d,
    eyeInnerColor: g
  } = n || {}, { path: l, logo: y } = v({ config: s, styleConfig: n, logoConfig: f, value: m }), { src: c, x: p, y: C, width: E, height: u, bg: e } = y || {};
  return /* @__PURE__ */ t.createElement(
    "svg",
    {
      width: r,
      height: r,
      viewBox: `0 0 ${r} ${r}`,
      style: {
        backgroundColor: i
      },
      className: "qr-container qr-svg" + (a ? ` ${a}` : "")
    },
    /* @__PURE__ */ t.createElement("g", { fill: o }, /* @__PURE__ */ t.createElement("path", { fill: d || o, d: l.eyeBorder }), /* @__PURE__ */ t.createElement("path", { fill: g || o, d: l.eyeInner }), /* @__PURE__ */ t.createElement("path", { d: l.points }), c && /* @__PURE__ */ t.createElement(t.Fragment, null, /* @__PURE__ */ t.createElement(
      "rect",
      {
        fill: i,
        x: e == null ? void 0 : e.x,
        y: e == null ? void 0 : e.y,
        width: e == null ? void 0 : e.width,
        height: e == null ? void 0 : e.height
      }
    ), /* @__PURE__ */ t.createElement("image", { href: c, width: E, height: u, x: p, y: C })))
  );
}
export {
  $ as default
};
