// amis-core/esm/actions/DrawerAction.js
/**
 * amis-core v6.13.0
 * Copyright 2018-2025 fex
 */

import { __awaiter, __generator, __assign } from 'tslib';
import 'amis-formula';
import 'moment';
import { extendObject } from '../utils/object.js';
import '../utils/memoryParse.js';
import '@rc-component/mini-decimal';
import 'lodash/isPlainObject';
import '../utils/filter.js';
import '../utils/api.js';
import 'file-saver';
import '../utils/browser.js';
import '../utils/ColorScale.js';
import 'react';
import 'lodash/chunk';
import '../utils/DataSchema.js';
import '../utils/DataScope.js';
import '../utils/debug.js';
import 'react-dom';
import '../utils/helper.js';
import '../utils/errors.js';
import '../utils/tpl.js';
import 'lodash/isObject';
import 'lodash/isString';
import 'lodash/isBoolean';
import '../utils/image.js';
import 'lodash/isEqual';
import { registerAction } from './Action.js';
import 'lodash/debounce';
import '../utils/resize-sensor.js';
import 'react-overlays/useRootClose';
import '../utils/SimpleMap.js';
import 'lodash/mapValues';
import 'lodash/camelCase';
import 'lodash/cloneDeep';
import 'lodash/map';
import 'lodash/isEmpty';
import 'lodash/kebabCase';
import 'uncontrollable';
import 'hoist-non-react-statics';
import '../utils/validations.js';
import '../utils/resolveCondition.js';
import 'mobx';
import '../utils/Animation.js';

/**
 * 打开抽屉动作
 *
 * @export
 * @class DrawerAction
 * @implements {Action}
 */
var DrawerAction = /** @class */ (function () {
    function DrawerAction() {
    }
    DrawerAction.prototype.run = function (action, renderer, event) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function () {
            var ret, _c, confirmed, value;
            var _d;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0:
                        // 防止editor preview模式下执行
                        if (action.$$id !== undefined) {
                            return [2 /*return*/];
                        }
                        ret = renderer.handleAction
                            ? renderer.handleAction(event, {
                                actionType: 'drawer',
                                drawer: action.drawer,
                                reload: 'none',
                                data: action.rawData
                            }, action.data)
                            : (_b = (_a = renderer.props).onAction) === null || _b === void 0 ? void 0 : _b.call(_a, event, {
                                actionType: 'drawer',
                                drawer: action.drawer,
                                reload: 'none',
                                data: action.rawData
                            }, action.data);
                        event.pendingPromise.push(ret);
                        if (!action.waitForAction) return [3 /*break*/, 2];
                        return [4 /*yield*/, ret];
                    case 1:
                        _c = _e.sent(), confirmed = _c.confirmed, value = _c.value;
                        event.setData(extendObject(event.data, (_d = {},
                            _d[action.outputVar || 'drawerResponse'] = {
                                confirmed: confirmed,
                                value: value
                            },
                            _d)));
                        _e.label = 2;
                    case 2: return [2 /*return*/];
                }
            });
        });
    };
    return DrawerAction;
}());
/**
 * 关闭抽屉动作
 *
 * @export
 * @class CloseDrawerAction
 * @implements {Action}
 */
var CloseDrawerAction = /** @class */ (function () {
    function CloseDrawerAction() {
    }
    CloseDrawerAction.prototype.run = function (action, renderer, event) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_c) {
                if (action.componentId) {
                    // 关闭指定抽屉
                    event.context.scoped.closeById(action.componentId);
                }
                else {
                    // 关闭当前抽屉
                    (_b = (_a = renderer.props).onAction) === null || _b === void 0 ? void 0 : _b.call(_a, event, __assign(__assign({}, action), { actionType: 'close' }), action.data);
                }
                return [2 /*return*/];
            });
        });
    };
    return CloseDrawerAction;
}());
registerAction('drawer', new DrawerAction());
registerAction('closeDrawer', new CloseDrawerAction());

export { CloseDrawerAction, DrawerAction };
