// office-viewer/lib/excel/render/ScrollBar.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var H = require('../../util/H.js');

/**
 * 模拟滚动条，目前暂时只支持浏览模式，也就是最大高宽是固定的
 */
var defaultOptions = {
    autoHide: true,
    scrollBarSize: 15
};
var ScrollBar = /** @class */ (function () {
    function ScrollBar(container, workbook, scrollOptions) {
        var _this = this;
        this.container = container;
        this.workbook = workbook;
        this.scrollOptions = tslib.__assign(tslib.__assign({}, defaultOptions), scrollOptions);
        workbook.uiEvent.on('AFTER_SCROLL', function () {
            _this.sync();
        });
        workbook.uiEvent.on('SWITCH_SHEET', function () {
            _this.sync();
        });
        var scrollBarSize = this.scrollOptions.scrollBarSize;
        container.style.position = 'relative';
        var yScrollBarContent = H.H('div', {
            className: 'ov-excel-scrollbar-content-y',
            style: {
                width: "".concat(scrollBarSize, "px")
            }
        });
        this.yScrollBarContent = yScrollBarContent;
        yScrollBarContent.addEventListener('scroll', function (event) {
            var currentSheet = _this.workbook.getActiveSheet();
            var sheetScrollTop = currentSheet.getScrollTop();
            var scrollBarTop = yScrollBarContent.scrollTop;
            if (Math.round(sheetScrollTop) !== Math.round(scrollBarTop)) {
                currentSheet.setScrollTop(yScrollBarContent.scrollTop);
                _this.workbook.uiEvent.emit('SCROLL_Y', yScrollBarContent.scrollTop);
            }
            event.preventDefault();
            event.stopPropagation();
        });
        // 避免误触发选中事件
        yScrollBarContent.addEventListener('mousedown', function (event) {
            event.stopPropagation();
        });
        var yScrollBarTotal = H.H('div', {
            className: 'ov-excel-scrollbar-total-y',
            style: {
                width: "".concat(scrollBarSize, "px")
            },
            parent: yScrollBarContent
        });
        this.yScrollBarTotal = yScrollBarTotal;
        var yScrollBarTruck = H.H('div', {
            className: 'ov-excel-scrollbar-y',
            style: {
                width: "".concat(scrollBarSize, "px")
            },
            parent: container
        }, yScrollBarContent);
        this.yScrollBarTruck = yScrollBarTruck;
        var xScrollBarContent = H.H('div', {
            className: 'ov-excel-scrollbar-content-x',
            style: {
                height: "".concat(scrollBarSize, "px")
            }
        });
        this.xScrollBarContent = xScrollBarContent;
        xScrollBarContent.addEventListener('scroll', function (event) {
            var currentSheet = _this.workbook.getActiveSheet();
            var sheetScrollLeft = currentSheet.getScrollLeft();
            var scrollBarLeft = xScrollBarContent.scrollLeft;
            if (Math.round(sheetScrollLeft) !== Math.round(scrollBarLeft)) {
                currentSheet.setScrollLeft(xScrollBarContent.scrollLeft);
                _this.workbook.uiEvent.emit('SCROLL_X', xScrollBarContent.scrollLeft);
            }
            event.preventDefault();
            event.stopPropagation();
        });
        // 避免误触发选中事件
        xScrollBarContent.addEventListener('mousedown', function (event) {
            event.stopPropagation();
        });
        var xScrollBarTotal = H.H('div', {
            className: 'ov-excel-scrollbar-total-x',
            style: {
                height: "".concat(scrollBarSize, "px")
            },
            parent: xScrollBarContent
        });
        this.xScrollBarTotal = xScrollBarTotal;
        var xScrollBarTruck = H.H('div', {
            className: 'ov-excel-scrollbar-x',
            style: {
                height: "".concat(scrollBarSize, "px")
            },
            parent: container
        }, xScrollBarContent);
        this.xScrollBarTruck = xScrollBarTruck;
        this.sync();
    }
    /**
     * 同步滚动条
     */
    ScrollBar.prototype.sync = function () {
        var currentSheet = this.workbook.getActiveSheet();
        var scrollBarSize = this.scrollOptions.scrollBarSize;
        var _a = this.workbook.getViewpointSize(), width = _a.width, height = _a.height;
        var scrollLeft = currentSheet.getScrollLeft();
        var scrollTop = currentSheet.getScrollTop();
        var _b = currentSheet.getRowColSize(), rowHeaderWidth = _b.rowHeaderWidth, colHeaderHeight = _b.colHeaderHeight;
        this.xScrollBarContent.style.width = "".concat(width, "px");
        this.yScrollBarContent.style.height = "".concat(height, "px");
        var totalWidth = Math.max(width + scrollLeft, currentSheet.getTotalWidth() + rowHeaderWidth + scrollBarSize);
        var totalHeight = Math.max(height + scrollTop, currentSheet.getTotalHeight() + colHeaderHeight + scrollBarSize);
        if (totalWidth <= width) {
            this.xScrollBarTruck.style.display = 'none';
        }
        else {
            this.xScrollBarTruck.style.display = 'block';
        }
        if (totalHeight <= height) {
            this.yScrollBarTruck.style.display = 'none';
        }
        else {
            this.yScrollBarTruck.style.display = 'block';
        }
        this.xScrollBarTotal.style.width = "".concat(totalWidth, "px");
        this.xScrollBarContent.scrollLeft = scrollLeft;
        this.yScrollBarTotal.style.height = "".concat(totalHeight, "px");
        this.yScrollBarContent.scrollTop = scrollTop;
    };
    return ScrollBar;
}());

exports.ScrollBar = ScrollBar;
