// amis-core/lib/StyleManager.js
/**
 * amis-core v6.13.0
 * Copyright 2018-2025 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var kebabCase = require('lodash/kebabCase');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var kebabCase__default = /*#__PURE__*/_interopDefaultLegacy(kebabCase);

var StyleManager = /** @class */ (function () {
    function StyleManager() {
        this.styles = {};
        this.styleDom = document.createElement('style');
        this.styleDom.id = 'amis-styles';
        document.head.appendChild(this.styleDom);
    }
    StyleManager.prototype.updateStyle = function (style) {
        var _this = this;
        Object.keys(style).forEach(function (className) {
            if (!_this.styles[className]) {
                _this.styles[className] = style[className];
            }
            else {
                _this.styles[className] = tslib.__assign(tslib.__assign({}, _this.styles[className]), style[className]);
            }
        });
        this.updateStyleDom();
    };
    StyleManager.prototype.removeStyles = function (id) {
        delete this.styles[id];
        this.updateStyleDom();
    };
    StyleManager.prototype.updateStyleDom = function () {
        var _this = this;
        var styleText = Object.keys(this.styles)
            .map(function (id) {
            var style = _this.styles[id];
            return Object.keys(style)
                .map(function (className) {
                return "".concat(className, " {").concat(Object.keys(style[className])
                    .map(function (propName) {
                    return "".concat(kebabCase__default["default"](propName), ": ").concat(style[className][propName], ";");
                })
                    .join(''), "}");
            })
                .join('');
        })
            .join('');
        this.styleDom.innerHTML = styleText;
        this.styleText = styleText;
    };
    return StyleManager;
}());
var styleManager = new StyleManager();

exports["default"] = styleManager;
