// amis-core/esm/utils/date.js
/**
 * amis-core v6.13.0
 * Copyright 2018-2025 fex
 */

import moment from 'moment';
import { createObject } from './object.js';
import { resolveVariableAndFilter } from './resolveVariableAndFilter.js';

var timeUnitMap = {
    year: 'Y',
    month: 'M',
    week: 'w',
    weekday: 'W',
    day: 'd',
    hour: 'h',
    minute: 'm',
    min: 'm',
    second: 's',
    millisecond: 'ms'
};
var relativeValueRe = /^(.+)?(\+|-)(\d+)(minute|min|hour|day|week|month|year|weekday|second|millisecond)s?$/i;
var filterDate = function (value, data, format, utc) {
    if (data === void 0) { data = {}; }
    if (format === void 0) { format = 'X'; }
    if (utc === void 0) { utc = false; }
    var m, mm = utc ? moment.utc : moment;
    if (typeof value === 'string') {
        value = value.trim();
    }
    // todo
    var date = new Date();
    value =
        typeof value === 'string'
            ? resolveVariableAndFilter(value, createObject(data, {
                now: mm().toDate(),
                today: mm([date.getFullYear(), date.getMonth(), date.getDate()])
            }), '| raw')
            : value;
    if (value && typeof value === 'string' && (m = relativeValueRe.exec(value))) {
        var date_1 = new Date();
        var step = parseInt(m[3], 10);
        var from = m[1]
            ? filterDate(m[1], data, format, utc)
            : mm(/(minute|min|hour|second)s?/.test(m[4])
                ? [
                    date_1.getFullYear(),
                    date_1.getMonth(),
                    date_1.getDate(),
                    date_1.getHours(),
                    date_1.getMinutes(),
                    date_1.getSeconds()
                ]
                : [date_1.getFullYear(), date_1.getMonth(), date_1.getDate()]);
        return m[2] === '-'
            ? from.subtract(step, timeUnitMap[m[4]])
            : from.add(step, timeUnitMap[m[4]]);
        //   return from[m[2] === '-' ? 'subtract' : 'add'](step, mapping[m[4]] || m[4]);
    }
    else if (value === 'now') {
        return mm();
    }
    else if (value === 'today') {
        var date_2 = new Date();
        return mm([date_2.getFullYear(), date_2.getMonth(), date_2.getDate()]);
    }
    else {
        // 优先通过指定格式解析，如果失败，则通过默认格式解析
        var date_3 = [mm(value, format), mm(value)].find(function (item) { return item.isValid(); }) || mm(value);
        return utc ? date_3.local() : date_3;
    }
};
function parseDuration(str) {
    var matches = /^((?:\-|\+)?(?:\d*\.)?\d+)(minute|min|hour|day|week|month|quarter|year|weekday|second|millisecond)s?$/.exec(str);
    if (matches) {
        var duration = moment.duration(parseFloat(matches[1]), matches[2]);
        if (moment.isDuration(duration)) {
            return duration;
        }
    }
    return;
}
/**
 * 解析日期，先尝试用 format 解析，如果失败，再尝试用其他标准格式解析
 * @param value
 * @param format
 * @returns
 */
function normalizeDate(value, format, options) {
    if (!value || value === '0') {
        return undefined;
    }
    var v = (options === null || options === void 0 ? void 0 : options.utc)
        ? moment.utc(value, format).local()
        : moment(value, format, true);
    if (v.isValid()) {
        return v;
    }
    if (typeof value === 'string' || typeof value === 'number') {
        var formats = ['', 'YYYY-MM-DD HH:mm:ss', 'X'];
        if (/^\d{10}((\.\d+)*)$/.test(value.toString())) {
            formats = ['X', 'x', 'YYYY-MM-DD HH:mm:ss', ''];
        }
        else if (/^\d{13}((\.\d+)*)$/.test(value.toString())) {
            formats = ['x', 'X', 'YYYY-MM-DD HH:mm:ss', ''];
        }
        while (formats.length) {
            var format_1 = formats.shift();
            var date = moment(value, format_1);
            if (date.isValid()) {
                return date;
            }
        }
    }
    return undefined;
}

export { filterDate, normalizeDate, parseDuration, relativeValueRe };
