// amis-core/esm/utils/debug.js
/**
 * amis-core v6.13.0
 * Copyright 2018-2025 fex
 */

import { __decorate, __metadata, __values, __read, __extends } from 'tslib';
import React, { useRef, useState, useEffect, Component } from 'react';
import cx from 'classnames';
import { findDOMNode, render, unmountComponentAtNode } from 'react-dom';
import { observable, action, autorun } from 'mobx';
import { observer } from 'mobx-react';
import { importLazyComponent, uuidv4 } from './helper.js';
import { position } from './position.js';
import { resolveVariableAndFilter } from './resolveVariableAndFilter.js';
import { callStrFunction } from './api.js';
import isPlainObject from 'lodash/isPlainObject';

/**
 * amis 运行时调试功能，为了避免循环引用，这个组件不要依赖 amis 里的组件
 */
var JsonView = React.lazy(function () {
    return import('react-json-view').then(importLazyComponent);
});
/** @class */ ((function () {
    function Log() {
        this.cat = '';
        this.level = '';
        this.msg = '';
        this.ext = '';
    }
    __decorate([
        observable,
        __metadata("design:type", Object)
    ], Log.prototype, "cat", void 0);
    __decorate([
        observable,
        __metadata("design:type", Object)
    ], Log.prototype, "level", void 0);
    __decorate([
        observable,
        __metadata("design:type", Object)
    ], Log.prototype, "msg", void 0);
    __decorate([
        observable,
        __metadata("design:type", Object)
    ], Log.prototype, "ext", void 0);
    return Log;
})());
var AMISDebugStore = /** @class */ (function () {
    function AMISDebugStore() {
        /**
         * 当前 tab
         */
        this.tab = 'log';
        /**
         * 显示位置，默认在右边
         */
        this.position = 'right';
        /**
         * 组件日志
         */
        this.logs = [];
        /**
         * Debug 面板是否展开
         */
        this.isExpanded = false;
        /**
         * 是否是 inspect 模式，在这个模式下可以查看数据域
         */
        this.inspectMode = false;
    }
    AMISDebugStore.prototype.open = function () {
        this.isExpanded = true;
        this.inspectMode = true;
    };
    AMISDebugStore.prototype.close = function () {
        this.isExpanded = false;
        this.activeId = '';
        this.hoverId = '';
        this.inspectMode = false;
    };
    AMISDebugStore.prototype.toggleInspectMode = function () {
        this.inspectMode = !this.inspectMode;
    };
    __decorate([
        observable,
        __metadata("design:type", String)
    ], AMISDebugStore.prototype, "tab", void 0);
    __decorate([
        observable,
        __metadata("design:type", String)
    ], AMISDebugStore.prototype, "position", void 0);
    __decorate([
        observable,
        __metadata("design:type", Array)
    ], AMISDebugStore.prototype, "logs", void 0);
    __decorate([
        observable,
        __metadata("design:type", Object)
    ], AMISDebugStore.prototype, "isExpanded", void 0);
    __decorate([
        observable,
        __metadata("design:type", Object)
    ], AMISDebugStore.prototype, "inspectMode", void 0);
    __decorate([
        observable,
        __metadata("design:type", String)
    ], AMISDebugStore.prototype, "hoverId", void 0);
    __decorate([
        observable,
        __metadata("design:type", String)
    ], AMISDebugStore.prototype, "activeId", void 0);
    __decorate([
        observable,
        __metadata("design:type", Number)
    ], AMISDebugStore.prototype, "ellipsisThreshold", void 0);
    __decorate([
        action.bound,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], AMISDebugStore.prototype, "open", null);
    __decorate([
        action.bound,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], AMISDebugStore.prototype, "close", null);
    __decorate([
        action.bound,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], AMISDebugStore.prototype, "toggleInspectMode", null);
    return AMISDebugStore;
}());
var store = new AMISDebugStore();
// 存储组件信息用于 debug
var ComponentInfo = {};
var LogView = observer(function (_a) {
    var _b;
    var store = _a.store;
    var logs = store.logs;
    var ellipsisThreshold = (_b = store.ellipsisThreshold) !== null && _b !== void 0 ? _b : 50;
    return (React.createElement(React.Fragment, null, logs.map(function (log, index) {
        var ext = typeof log.ext === 'string' &&
            (log.ext.startsWith('{') || log.ext.startsWith('['))
            ? parseJson(log.ext)
            : typeof log.ext === 'object'
                ? normalizeDataForLog(log.ext)
                : log.ext;
        return (React.createElement("div", { className: "AMISDebug-logLine", key: "log-".concat(index) },
            React.createElement("div", { className: "AMISDebug-logLineMsg" },
                "[",
                log.cat,
                "] ",
                log.msg),
            typeof ext === 'object' ? (React.createElement(React.Suspense, { fallback: React.createElement("div", null, "Loading...") },
                React.createElement(JsonView, { name: null, theme: "monokai", src: ext, collapsed: true, enableClipboard: false, displayDataTypes: false, collapseStringsAfterLength: ellipsisThreshold, iconStyle: "square" }))) : (React.createElement("pre", { className: "AMISDebug-value" }, JSON.stringify(ext)))));
    })));
});
var AMISDebug = observer(function (_a) {
    var e_1, _b;
    var _c, _d;
    var store = _a.store;
    var activeId = store.activeId;
    var activeComponentInspect = ComponentInfo[activeId];
    // 收集数据域里的数据
    var start = ((_d = (_c = activeComponentInspect === null || activeComponentInspect === void 0 ? void 0 : activeComponentInspect.component) === null || _c === void 0 ? void 0 : _c.props) === null || _d === void 0 ? void 0 : _d.data) || {};
    var stacks = [start];
    while (Object.getPrototypeOf(start) !== Object.prototype) {
        var superData = Object.getPrototypeOf(start);
        if (Object.prototype.toString.call(superData) !== '[object Object]') {
            break;
        }
        stacks.push(superData);
        start = superData;
    }
    var stackDataView = [];
    if (Object.keys(stacks[0]).length || stacks.length > 1) {
        var level = 0;
        try {
            for (var stacks_1 = __values(stacks), stacks_1_1 = stacks_1.next(); !stacks_1_1.done; stacks_1_1 = stacks_1.next()) {
                var stack = stacks_1_1.value;
                stackDataView.push(React.createElement("div", { key: "data-".concat(level) },
                    React.createElement("h3", null,
                        "Data Level-",
                        level),
                    React.createElement(React.Suspense, { fallback: React.createElement("div", null, "Loading...") },
                        React.createElement(JsonView, { key: "dataview-".concat(stack), name: null, theme: "monokai", src: stack, collapsed: level === 0 ? false : true, enableClipboard: false, displayDataTypes: false, iconStyle: "square" }))));
                level += 1;
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (stacks_1_1 && !stacks_1_1.done && (_b = stacks_1.return)) _b.call(stacks_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
    }
    var panelRef = useRef(null);
    var _e = __read(useState(false), 2), isResizing = _e[0], setResizing = _e[1];
    var _f = __read(useState(0), 2), startX = _f[0], setStartX = _f[1];
    var _g = __read(useState(0), 2), panelWidth = _g[0], setPanelWidth = _g[1];
    useEffect(function () {
        var handleMouseUp = function () {
            setResizing(false);
        };
        var handleMouseMove = function (e) {
            if (!isResizing) {
                return;
            }
            var xOffset = store.position === 'right' ? e.clientX - startX : startX - e.clientX;
            var panel = panelRef.current;
            var targetWidth = Math.max(200, panelWidth - xOffset);
            panel.style.width = targetWidth + 'px';
            if (e.stopPropagation)
                e.stopPropagation();
            if (e.preventDefault)
                e.preventDefault();
            e.cancelBubble = true;
            return false;
        };
        if (isResizing) {
            document.addEventListener('mousemove', handleMouseMove);
            document.addEventListener('mouseup', handleMouseUp);
        }
        return function () {
            if (isResizing) {
                document.removeEventListener('mousemove', handleMouseMove);
                document.removeEventListener('mouseup', handleMouseUp);
            }
        };
    }, [isResizing]);
    // 避免触发 modal 的 closeOnOutside 逻辑
    useEffect(function () {
        var handlePanelMouseUp = function (e) {
            e.preventDefault();
        };
        panelRef.current.addEventListener('mouseup', handlePanelMouseUp);
        return function () {
            var _a;
            (_a = panelRef.current) === null || _a === void 0 ? void 0 : _a.removeEventListener('mouseup', handlePanelMouseUp);
        };
    }, []);
    var handleInputKeyUp = React.useCallback(function (e) {
        var _a;
        if (e.key === 'Enter') {
            var input = e.target;
            var expression = input.value;
            input.value = '';
            try {
                var activeId_1 = store.activeId;
                var ctx = ((_a = ComponentInfo[activeId_1]) === null || _a === void 0 ? void 0 : _a.component.props.data) || {};
                var result = expression.startsWith('${')
                    ? resolveVariableAndFilter(expression, ctx, '| raw')
                    : callStrFunction(new Function('data', "with(data) {return ".concat(expression, ";}")), ['data'], ctx);
                debug('debug', "evaluate expression `".concat(expression, "`"), result);
            }
            catch (e) {
                debug('error', "evaluate expression `".concat(expression, "`"), "Error: ".concat(e.message));
            }
            finally {
                store.tab = 'log';
            }
        }
    }, []);
    return (React.createElement("div", { className: cx('AMISDebug', {
            'is-expanded': store.isExpanded,
            'is-left': store.position === 'left'
        }), ref: panelRef },
        React.createElement("div", { className: "AMISDebug-toggle", title: "open debug", onClick: store.open }, store.isExpanded ? (React.createElement("i", { className: "fas fa-times" })) : (React.createElement("i", { className: "fas fa-bug" }))),
        store.isExpanded ? (React.createElement("div", { className: cx('AMISDebug-content') },
            React.createElement("div", { className: "AMISDebug-close", title: "Close", onClick: store.close },
                React.createElement("i", { className: "fas fa-times" })),
            React.createElement("div", { className: "AMISDebug-resize", onMouseDown: function (event) {
                    setStartX(event.clientX);
                    setPanelWidth(parseInt(getComputedStyle(panelRef.current).getPropertyValue('width'), 10));
                    setResizing(true);
                } }),
            React.createElement("div", { className: "AMISDebug-tab" },
                React.createElement("button", { className: cx({ active: store.tab === 'log' }), onClick: function () {
                        store.tab = 'log';
                    } },
                    "Log(",
                    store.logs.length,
                    ")"),
                React.createElement("button", { className: cx({ active: store.tab === 'inspect' }), onClick: function () {
                        store.tab = 'inspect';
                    } }, "Inspect")),
            React.createElement("div", { className: "AMISDebug-changePosition" }, store.position === 'right' ? (React.createElement("i", { className: "fas fa-chevron-left", title: "move to left", onClick: function () {
                    store.position = 'left';
                } })) : (React.createElement("i", { className: "fas fa-chevron-right", title: "move to right", onClick: function () {
                    store.position = 'right';
                } }))),
            store.tab === 'log' ? (React.createElement("div", { className: "AMISDebug-log" },
                React.createElement("button", { onClick: function () {
                        store.logs = [];
                    } }, "Clear Log"),
                React.createElement(LogView, { store: store }))) : null,
            store.tab === 'inspect' ? (React.createElement("div", { className: "AMISDebug-inspect" },
                React.createElement("svg", { xmlns: "http://www.w3.org/2000/svg", fill: "currentColor", viewBox: "0 0 1024 1024", className: "AMISDebug-inspectIcon ".concat(store.inspectMode ? 'is-active' : ''), onClick: store.toggleInspectMode },
                    React.createElement("path", { d: "M64 192L128 128h768l64 64v384H896v-384H128v512h320V768H128l-64-64v-512z m941.226667 621.226667L576 384v602.496l173.226667-173.226667h256zM640 832v-293.504l210.773333 210.773333h-128L640 832z" })),
                store.inspectMode ? (React.createElement("span", null, "Select an element in the page to inspect it.")) : (React.createElement("span", null, "Click to inspect an element.")),
                activeId ? (React.createElement(React.Fragment, null,
                    React.createElement("h3", null,
                        "Component:",
                        ' ',
                        React.createElement("span", { className: "primary" }, activeComponentInspect === null || activeComponentInspect === void 0 ? void 0 : activeComponentInspect.name)),
                    stackDataView)) : null)) : null,
            activeId ? (React.createElement("div", { className: "AMISDebug-footer" },
                React.createElement("svg", { xmlns: "http://www.w3.org/2000/svg", fill: "currentColor", viewBox: "0 0 1024 1024" },
                    React.createElement("path", { d: "M724.48 521.728c-1.8432 7.7824-5.7344 14.848-11.3664 20.48l-341.9136 342.016c-16.6912 16.6912-43.7248 16.6912-60.3136 0s-16.6912-43.7248 0-60.3136L622.6944 512 310.8864 200.0896c-16.6912-16.6912-16.6912-43.7248 0-60.3136 16.6912-16.6912 43.7248-16.6912 60.3136 0l341.9136 341.9136c10.8544 10.8544 14.6432 26.112 11.3664 40.0384z", fill: "currentColor" })),
                React.createElement("input", { type: "text", placeholder: "", onKeyUp: handleInputKeyUp }))) : null)) : null));
});
/**
 * 鼠标移动到某个组件的效果
 */
function handleMouseMove(e) {
    if (!store.inspectMode) {
        return;
    }
    var dom = e.target;
    var target = dom.closest("[data-debug-id]");
    if (target) {
        store.hoverId = target.getAttribute('data-debug-id');
    }
}
/**
 *  点选某个组件
 */
function handleMouseclick(e) {
    if (!store.inspectMode) {
        return;
    }
    e.preventDefault();
    var dom = e.target;
    var target = dom.closest("[data-debug-id]");
    if (target && !target.closest('.AMISDebug')) {
        store.activeId = target.getAttribute('data-debug-id');
        store.tab = 'inspect';
        store.inspectMode = false;
    }
}
// hover 及点击后的高亮
var amisHoverBox = document.createElement('div');
amisHoverBox.className = 'AMISDebug-hoverBox';
var amisActiveBox = document.createElement('div');
amisActiveBox.className = 'AMISDebug-activeBox';
var timer = null;
autorun(function () {
    var hoverId = store.hoverId;
    var hoverElement = document.querySelector("[data-debug-id=\"".concat(hoverId, "\"]"));
    if (hoverElement) {
        var offset = position(hoverElement, document.body);
        amisHoverBox.style.top = "".concat(offset.top, "px");
        amisHoverBox.style.left = "".concat(offset.left, "px");
        amisHoverBox.style.width = "".concat(offset.width, "px");
        amisHoverBox.style.height = "".concat(offset.height, "px");
    }
    else {
        amisHoverBox.style.top = '-999999px';
    }
});
autorun(function () {
    var activeId = store.activeId;
    var activeElement = document.querySelector("[data-debug-id=\"".concat(activeId, "\"]"));
    if (activeElement) {
        var offset = position(activeElement, document.body);
        amisActiveBox.style.top = "".concat(offset.top, "px");
        amisActiveBox.style.left = "".concat(offset.left, "px");
        amisActiveBox.style.width = "".concat(offset.width, "px");
        amisActiveBox.style.height = "".concat(offset.height, "px");
        amisActiveBox.classList.add('shake');
        clearTimeout(timer);
        timer = setTimeout(function () {
            amisActiveBox.classList.remove('shake');
        }, 500);
    }
    else {
        amisActiveBox.style.top = '-999999px';
    }
});
// 页面中只能有一个实例
var isEnabled = false;
var unmount;
function enableDebug() {
    if (isEnabled) {
        return;
    }
    isEnabled = true;
    var amisDebugElement = document.createElement('div');
    document.body.appendChild(amisDebugElement);
    var element = React.createElement(AMISDebug, { store: store });
    // if (parseInt(version.split('.')[0], 10) >= 18) {
    //   const root = createRoot(amisDebugElement);
    //   root.render(element);
    //   unmount = () => {
    //     root.unmount();
    //     document.body.removeChild(amisDebugElement);
    //   };
    // } else {
    render(element, amisDebugElement);
    unmount = function () {
        unmountComponentAtNode(amisDebugElement);
        document.body.removeChild(amisDebugElement);
    };
    // }
    document.body.appendChild(amisHoverBox);
    document.body.appendChild(amisActiveBox);
    document.addEventListener('mousemove', handleMouseMove);
    document.addEventListener('click', handleMouseclick, true);
}
function disableDebug() {
    if (!isEnabled) {
        return;
    }
    isEnabled = false;
    unmount === null || unmount === void 0 ? void 0 : unmount();
    document.body.removeChild(amisHoverBox);
    document.body.removeChild(amisActiveBox);
    document.removeEventListener('mousemove', handleMouseMove);
    document.removeEventListener('click', handleMouseclick);
}
var DebugWrapper = /** @class */ (function (_super) {
    __extends(DebugWrapper, _super);
    function DebugWrapper() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.debugId = uuidv4();
        return _this;
    }
    DebugWrapper.prototype.componentDidMount = function () {
        var root = findDOMNode(this);
        if (!root) {
            return;
        }
        var renderer = this.props.renderer;
        root.setAttribute('data-debug-id', this.debugId);
        ComponentInfo[this.debugId] = {
            name: renderer.name,
            component: this.props.children
        };
    };
    DebugWrapper.prototype.componentDidUpdate = function (prevProps) {
        var renderer = this.props.renderer;
        if (!ComponentInfo[this.debugId]) {
            return;
        }
        ComponentInfo[this.debugId] = {
            name: renderer.name,
            component: this.props.children
        };
    };
    DebugWrapper.prototype.componentWillUnmount = function () {
        delete ComponentInfo[this.debugId];
        if (this.debugId === store.activeId) {
            store.activeId = '';
        }
        if (this.debugId === store.hoverId) {
            store.hoverId = '';
        }
    };
    DebugWrapper.prototype.render = function () {
        return this.props.children;
    };
    return DebugWrapper;
}(Component));
function parseJson(str) {
    try {
        return JSON.parse(str);
    }
    catch (e) {
        return e;
    }
}
function normalizeDataForLog(data) {
    try {
        return parseJson(JSON.stringify(data));
    }
    catch (_a) {
        var ret_1 = {};
        Object.keys(data).forEach(function (key) {
            var value = data[key];
            if (typeof value === 'object' &&
                value !== null &&
                !isPlainObject(value)) {
                ret_1[key] = '[complicated data]';
            }
            else {
                ret_1[key] = value;
            }
        });
        return ret_1;
    }
}
function safeStringify(data) {
    try {
        return JSON.stringify(data);
    }
    catch (_a) {
        return data;
    }
}
/**
 * 一般调试日志
 * @param msg 简单消息
 * @param ext 扩展信息
 */
function debug(cat, msg, ext) {
    if (!isEnabled) {
        return;
    }
    console.groupCollapsed('[amis debug]', msg);
    console.debug(ext);
    console.groupEnd();
    var log = {
        cat: cat,
        level: 'debug',
        msg: msg,
        ext: safeStringify(ext) // 因为 mobx 会加工，导致性能问题，所以转成字符串最简单
    };
    store.logs.push(log);
    // 不要超过 200 条，担心性能问题
    if (store.logs.length > 200) {
        store.logs.splice(0, store.logs.length - 200);
    }
}
/**
 * 警告日志
 * @param msg 简单消息
 * @param ext 扩展信息
 */
function warning(cat, msg, ext) {
    if (!isEnabled) {
        return;
    }
    var log = {
        cat: cat,
        level: 'warn',
        msg: msg,
        ext: ext
    };
    console.groupCollapsed('amis debug', msg);
    console.trace(log);
    console.groupEnd();
    store.logs.push(log);
}
// 辅助定位是因为什么属性变化导致了组件更新
function traceProps(props, prevProps, componentName) {
    console.log(componentName, Object.keys(props)
        .map(function (key) {
        if (props[key] !== prevProps[key]) {
            if (key === 'data') {
                return "data[".concat(Object.keys(props[key])
                    .map(function (item) {
                    if (props[key][item] !== prevProps[key][item]) {
                        return "".concat(item);
                    }
                    return '';
                })
                    .filter(function (item) { return item; })
                    .join(', '), "]");
            }
            return key;
        }
        return '';
    })
        .filter(function (item) { return item; }));
}

export { DebugWrapper, JsonView, debug, disableDebug, enableDebug, safeStringify, traceProps, warning };
