// amis-core/esm/utils/resize-sensor.js
/**
 * amis-core v6.13.0
 * Copyright 2018-2025 fex
 */

/**
 * @file resize-sensor.js.
 * @author fex
 */
var EventQueue = /** @class */ (function () {
    function EventQueue() {
        this.q = [];
    }
    EventQueue.prototype.add = function (cb, type) {
        if (type === void 0) { type = 'both'; }
        this.q.push({
            fn: cb,
            type: type
        });
    };
    EventQueue.prototype.call = function (type) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        this.q.forEach(function (item) {
            if (item.type === type || item.type === 'both' || type === 'both') {
                item.fn.apply(null, args);
            }
        });
    };
    return EventQueue;
}());
function getComputedStyle(element, prop) {
    if (element.currentStyle) {
        return element.currentStyle[prop];
    }
    else if (window.getComputedStyle) {
        var style = window.getComputedStyle(element, undefined);
        return style ? style.getPropertyValue(prop) : undefined;
    }
    else {
        return element.style[prop];
    }
}
function attachResizeEvent(element, resized, type) {
    if (type === void 0) { type = 'both'; }
    if (!element) {
        return;
    }
    if (!element.resizedAttached) {
        element.resizedAttached = new EventQueue();
        element.resizedAttached.add(resized, type);
    }
    else if (element.resizedAttached) {
        element.resizedAttached.add(resized, type);
        return;
    }
    var resizeSensor = (element.resizeSensor =
        document.createElement('div'));
    resizeSensor.className = 'resize-sensor';
    var style = 'position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;';
    var styleChild = 'position: absolute; left: 0; top: 0;';
    resizeSensor.style.cssText = style;
    resizeSensor.innerHTML = "\n  <div class=\"resize-sensor-expand\" style=\"".concat(style, "\">\n    <div style=\"").concat(styleChild, "\"></div>\n  </div>\n  <div class=\"resize-sensor-shrink\" style=\"").concat(style, "\">\n    <div style=\"").concat(styleChild, " width: 200%; height: 200%\"></div>\n  </div>\n  <div class=\"resize-sensor-appear\" style=\"").concat(style, "animation-name: apearSensor; animation-duration: 0.2s;\"></div>");
    // 要定义 resizeSensor 这个动画，靠这个监听出现。
    element.appendChild(resizeSensor);
    element.hasInlineStyle = element.hasAttribute('style');
    var position = (element.originPosition = getComputedStyle(element, 'position'));
    if (!~['fixed', 'absolute'].indexOf(position)) {
        element.style.position = 'relative';
    }
    var expand = resizeSensor.children[0];
    var expandChild = expand.children[0];
    var shrink = resizeSensor.children[1];
    // let shrinkChild = shrink.children[0] as HTMLElement;
    var appear = resizeSensor.children[2];
    var lastWidth, lastHeight;
    var reset = function () {
        expandChild.style.width = expand.offsetWidth + 10 + 'px';
        expandChild.style.height = expand.offsetHeight + 10 + 'px';
        expand.scrollLeft = expand.scrollWidth;
        expand.scrollTop = expand.scrollHeight;
        shrink.scrollLeft = shrink.scrollWidth;
        shrink.scrollTop = shrink.scrollHeight;
        lastWidth = element.offsetWidth;
        lastHeight = element.offsetHeight;
    };
    var appeared = function () {
        reset();
        // 如果初始是隐藏状态，首次显示也触发resize事件。
        if (isHidden) {
            changed();
            isHidden = false;
        }
    };
    reset();
    var changed = function (type) {
        if (type === void 0) { type = 'both'; }
        if (element.resizedAttached) {
            element.resizedAttached.call(type);
        }
    };
    var addEvent = function (el, name, cb) {
        if (el.attachEvent) {
            el.attachEvent('on' + name, cb);
        }
        else {
            el.addEventListener(name, cb);
        }
    };
    var removeEvent = function (el, name, cb) {
        if (el.detachEvent) {
            el.detachEvent('on' + name, cb);
        }
        else {
            el.removeEventListener(name, cb);
        }
    };
    var onScroll = function (e) {
        var widthChanged = element.offsetWidth != lastWidth;
        var heightChanged = element.offsetHeight != lastHeight;
        if (widthChanged || heightChanged) {
            changed(widthChanged && heightChanged
                ? 'both'
                : widthChanged
                    ? 'width'
                    : 'height');
        }
        reset();
    };
    addEvent(expand, 'scroll', onScroll);
    addEvent(shrink, 'scroll', onScroll);
    addEvent(appear, 'animationstart', appeared);
    var isHidden = !expand.offsetWidth;
    return function () {
        removeEvent(expand, 'scroll', onScroll);
        removeEvent(shrink, 'scroll', onScroll);
        removeEvent(appear, 'animationstart', appeared);
    };
}
function detach(element) {
    if (element.resizeSensor) {
        if (element.hasInlineStyle) {
            element.style.position = element.originPosition;
        }
        else {
            element.removeAttribute('style');
        }
        try {
            element.removeChild(element.resizeSensor);
        }
        catch (e) { }
        delete element.resizeSensor;
        delete element.resizedAttached;
        delete element.hasInlineStyle;
        delete element.originPosition;
    }
}
function appearSensor(element, callback, once) {
    if (once === void 0) { once = false; }
    if (once) {
        callback = (function (original) {
            return function () {
                original();
                dispose();
            };
        })(callback);
    }
    element.addEventListener('animationstart', callback);
    var originalAnimationName = element.style.animationName;
    var originalAnimationDuration = element.style.animationDuration;
    element.style.cssText =
        'animation-name: apearSensor; animation-duration: 0.2s;';
    var dispose = function () {
        element.style.animationName = originalAnimationName;
        element.style.animationDuration = originalAnimationDuration;
        element.removeEventListener('animationstart', callback);
    };
    return dispose;
}
function resizeSensorV2(element, callback, once, type) {
    if (once === void 0) { once = false; }
    if (type === void 0) { type = 'both'; }
    var rect = element.getBoundingClientRect();
    var originWidth = rect.width;
    var originHeight = rect.height;
    var observer = new ResizeObserver(function (entries) {
        if (once) {
            observer.disconnect();
        }
        var entry = entries[0];
        var cr = entry.contentRect;
        // 变化大于0.5px时才触发回调,允许一定的误差
        var widthChanged = Math.abs(cr.width - originWidth) > 0.5;
        var heightChanged = Math.abs(cr.height - originHeight) > 0.5;
        if (widthChanged || heightChanged) {
            if (type === 'both') {
                callback();
            }
            else if ((type === 'width' && widthChanged) ||
                (type === 'height' && heightChanged)) {
                callback();
            }
            originWidth = cr.width;
            originHeight = cr.height;
        }
    });
    observer.observe(element);
    return function () {
        observer.disconnect();
    };
}
function resizeSensor(element, callback, once, type, triggerOnAppear) {
    if (once === void 0) { once = false; }
    if (type === void 0) { type = 'both'; }
    if (triggerOnAppear === void 0) { triggerOnAppear = false; }
    if (!element) {
        return function () { };
    }
    // 优先用 ResizeObserver
    if (typeof ResizeObserver !== 'undefined') {
        var disposes_1 = [
            resizeSensorV2(element, callback, once, type),
            triggerOnAppear ? appearSensor(element, callback, true) : undefined
        ];
        return function () {
            disposes_1.forEach(function (dispose) { return dispose === null || dispose === void 0 ? void 0 : dispose(); });
        };
    }
    // 不支持 ResizeObserver 的话，用 polyfill
    var disposeEvent = undefined;
    if (once) {
        disposeEvent = attachResizeEvent(element, function () {
            callback.apply(this, arguments);
            disposeEvent === null || disposeEvent === void 0 ? void 0 : disposeEvent();
            detach(element);
        }, type);
        return function () { };
    }
    disposeEvent = attachResizeEvent(element, callback, type);
    var detached = false;
    return function () {
        if (detached)
            return;
        detached = true;
        disposeEvent === null || disposeEvent === void 0 ? void 0 : disposeEvent();
        detach(element);
    };
}

export { appearSensor, getComputedStyle, resizeSensor, resizeSensorV2 };
