// amis-core/esm/utils/tpl-builtin.js
/**
 * amis-core v6.13.0
 * Copyright 2018-2025 fex
 */

import { __awaiter, __generator } from 'tslib';
export { getFilters, registerFilter, registerFunction } from 'amis-formula';
import 'moment';
import 'lodash/isPlainObject';
import './memoryParse.js';
import { resolveVariableAndFilter } from './resolveVariableAndFilter.js';
export { resolveVariableAndFilter } from './resolveVariableAndFilter.js';
import '@rc-component/mini-decimal';
import { tokenize } from './tokenize.js';
export { tokenize } from './tokenize.js';
import { resolveVariableAndFilterForAsync } from './resolveVariableAndFilterForAsync.js';
export { resolveVariableAndFilterForAsync } from './resolveVariableAndFilterForAsync.js';
import './filter.js';

function matchSynatax(str) {
    var from = 0;
    while (true) {
        var idx = str.indexOf('$', from);
        if (~idx) {
            var nextToken = str[idx + 1];
            // 如果没有下一个字符，或者下一个字符是引号或者空格
            // 这个一般不是取值用法
            if (!nextToken || ~['"', "'", ' '].indexOf(nextToken)) {
                from = idx + 1;
                continue;
            }
            // 如果上个字符是转义也不是取值用法
            var prevToken = str[idx - 1];
            if (prevToken && prevToken === '\\') {
                from = idx + 1;
                continue;
            }
            return true;
        }
        else {
            break;
        }
    }
    return false;
}
function register() {
    var _this = this;
    return {
        name: 'builtin',
        test: function (str) { return typeof str === 'string' && matchSynatax(str); },
        removeEscapeToken: function (str) {
            return typeof str === 'string' ? str.replace(/\\\$/g, '$') : str;
        },
        compile: function (str, data, defaultFilter) {
            if (defaultFilter === void 0) { defaultFilter = '| html'; }
            try {
                return tokenize(str, data, defaultFilter);
            }
            catch (e) {
                return "error: ".concat(e.message);
            }
        },
        asyncCompile: function (str, data, defaultFilter) {
            if (defaultFilter === void 0) { defaultFilter = '| html'; }
            return __awaiter(_this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    try {
                        return [2 /*return*/, resolveVariableAndFilterForAsync(str, data, defaultFilter)];
                    }
                    catch (e) {
                        return [2 /*return*/, "error: ".concat(e.message)];
                    }
                    return [2 /*return*/];
                });
            });
        }
    };
}
// 避免 resolveVariableAndFilter 被摇了
var testResolveVariableAndFilter = resolveVariableAndFilter;

export { register, testResolveVariableAndFilter };
