// office-viewer/esm/excel/render/drawing/chart/fromLineChart.js
import { __values } from 'tslib';
import { calcPercentStacked } from './calcPercentStacked.js';
import { getStack } from './getStack.js';
import { getData } from './getData.js';
import { buildLabel } from './buildLabel.js';

function fromLineChart(workbook, lineChart) {
    var e_1, _a;
    var _b, _c, _d, _e, _f;
    var categories = [];
    var series = [];
    var ser = lineChart.ser || [];
    var isPercentStacked = false;
    try {
        for (var ser_1 = __values(ser), ser_1_1 = ser_1.next(); !ser_1_1.done; ser_1_1 = ser_1.next()) {
            var lineSeries = ser_1_1.value;
            var seriesData = getData(workbook, lineSeries.val);
            var name_1 = ((_f = (_e = (_d = (_c = (_b = lineSeries.tx) === null || _b === void 0 ? void 0 : _b.strRef) === null || _c === void 0 ? void 0 : _c.strCache) === null || _d === void 0 ? void 0 : _d.pt) === null || _e === void 0 ? void 0 : _e[0]) === null || _f === void 0 ? void 0 : _f.v) || '';
            if (name_1) {
                categories.push(name_1);
            }
            var label = buildLabel(lineSeries.dLbls);
            var stack = getStack(lineChart.grouping);
            isPercentStacked = stack.isPercentStacked;
            series.push({
                name: name_1,
                type: 'line',
                stack: stack.stack,
                data: seriesData,
                label: label
            });
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (ser_1_1 && !ser_1_1.done && (_a = ser_1.return)) _a.call(ser_1);
        }
        finally { if (e_1) throw e_1.error; }
    }
    if (isPercentStacked) {
        calcPercentStacked(series);
    }
    return {
        categories: categories,
        series: series
    };
}

export { fromLineChart };
