// office-viewer/esm/excel/render/sheetTab/StatusBar.js
import { __values } from 'tslib';
import { H } from '../../../util/H.js';
import { stripNumber } from '../../../util/stripNumber.js';

/**
 * 状态栏，主要是显示平均值、求和等
 */
var StatusBar = /** @class */ (function () {
    function StatusBar(container, workbook, renderOptions) {
        var _this = this;
        this.workbook = workbook;
        var status = H('div', {
            className: 'ov-excel-sheet-tab-bar__status',
            parent: container
        });
        this.status = status;
        workbook.uiEvent.on('CHANGE_SELECTION', function () {
            _this.syncStatus();
        });
    }
    /**
     * 选区变化的时候更新底部状态栏
     */
    StatusBar.prototype.syncStatus = function () {
        var e_1, _a;
        var workbook = this.workbook;
        var status = this.status;
        var currentSheet = workbook.getActiveSheet();
        var selection = currentSheet.getSelection();
        if (!selection) {
            status.textContent = '';
            return;
        }
        var rangesData = currentSheet.getCellValueByRanges(selection.cellRanges, false);
        var count = rangesData.length;
        var sum = 0;
        var numberCount = 0;
        try {
            for (var rangesData_1 = __values(rangesData), rangesData_1_1 = rangesData_1.next(); !rangesData_1_1.done; rangesData_1_1 = rangesData_1.next()) {
                var data = rangesData_1_1.value;
                if (data) {
                    if (data.isDate) {
                        continue;
                    }
                    var numberVal = parseFloat((data.value || data.text) + ''.trim());
                    if (isNaN(numberVal)) {
                        continue;
                    }
                    sum += numberVal;
                    numberCount++;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (rangesData_1_1 && !rangesData_1_1.done && (_a = rangesData_1.return)) _a.call(rangesData_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        sum = stripNumber(sum, 10);
        var average = numberCount ? stripNumber(sum / numberCount, 10) : 0;
        status.innerHTML = '';
        H('span', {
            class: 'ov-excel-sheet-tab-bar__status-text',
            parent: status,
            innerText: "".concat(workbook.translator('average'), ": ").concat(average)
        });
        H('span', {
            class: 'ov-excel-sheet-tab-bar__status-text',
            parent: status,
            innerText: "".concat(workbook.translator('count'), ": ").concat(count)
        });
        H('span', {
            class: 'ov-excel-sheet-tab-bar__status-text',
            parent: status,
            innerText: "".concat(workbook.translator('sum'), ": ").concat(sum)
        });
    };
    return StatusBar;
}());

export { StatusBar };
