// office-viewer/lib/openxml/word/Paragraph.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var OpenXML = require('../../OpenXML.js');
var parsePr = require('../../word/parse/parsePr.js');
var Bookmark = require('./Bookmark.js');
var Hyperlink = require('./Hyperlink.js');
var NumberProperties = require('./numbering/NumberProperties.js');
var Run = require('./Run.js');
var Tab = require('./Tab.js');
var FldSimple = require('./FldSimple.js');
var OMath = require('../math/OMath.js');
var mergeSdt = require('../../word/parse/mergeSdt.js');

/**
 * 段落的定义和解析
 */
// | SymbolRun
// | PageBreak
// | ColumnBreak
// | SequentialIdentifier
// | FootnoteReferenceRun
// | InsertedTextRun
// | DeletedTextRun
// | Math
// | SimpleField
// | SimpleMailMergeField
// | Comments
// | Comment
// | CommentRangeStart
// | CommentRangeEnd
// | CommentReference;
function parseAutoSpace(element) {
    var autoSpaceDE = element.getElementsByTagName('w:autoSpaceDE').item(0);
    var autoSpaceDN = element.getElementsByTagName('w:autoSpaceDN').item(0);
    return !!autoSpaceDE || !!autoSpaceDN;
}
var Paragraph = /** @class */ (function () {
    function Paragraph() {
        this.properties = {};
        this.children = [];
        this.fldSimples = [];
    }
    Paragraph.prototype.addChild = function (child) {
        this.children.push(child);
    };
    Paragraph.parseParagraphPr = function (word, element) {
        var e_1, _a;
        var cssStyle = parsePr.parsePr(word, element);
        var pStyle;
        var pStyleTag = element.getElementsByTagName('w:pStyle').item(0);
        if (pStyleTag) {
            pStyle = OpenXML.getVal(pStyleTag);
        }
        var numPr;
        var numPrTag = element.getElementsByTagName('w:numPr').item(0);
        if (numPrTag) {
            numPr = NumberProperties.NumberPr.fromXML(word, numPrTag);
        }
        var tabs = [];
        var tabElements = element.getElementsByTagName('w:tab');
        try {
            for (var tabElements_1 = tslib.__values(tabElements), tabElements_1_1 = tabElements_1.next(); !tabElements_1_1.done; tabElements_1_1 = tabElements_1.next()) {
                var tabElement = tabElements_1_1.value;
                tabs.push(Tab.Tab.fromXML(word, tabElement));
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (tabElements_1_1 && !tabElements_1_1.done && (_a = tabElements_1.return)) _a.call(tabElements_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        var autoSpace = parseAutoSpace(element);
        return { cssStyle: cssStyle, pStyle: pStyle, numPr: numPr, tabs: tabs, autoSpace: autoSpace };
    };
    Paragraph.fromXML = function (word, element) {
        var e_2, _a;
        var paragraph = new Paragraph();
        paragraph.fldSimples = [];
        paragraph.paraId = element.getAttribute('w14:paraId') || '';
        try {
            for (var _b = tslib.__values(mergeSdt.mergeSdt(element)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var child = _c.value;
                var tagName = child.tagName;
                switch (tagName) {
                    case 'w:pPr':
                        paragraph.properties = Paragraph.parseParagraphPr(word, child);
                        break;
                    case 'w:r':
                        paragraph.addChild(Run.Run.fromXML(word, child));
                        break;
                    case 'w:hyperlink':
                        paragraph.addChild(Hyperlink.Hyperlink.fromXML(word, child));
                        break;
                    case 'w:bookmarkStart':
                        paragraph.addChild(Bookmark.BookmarkStart.fromXML(word, child));
                    case 'w:bookmarkEnd':
                        // 没啥用所以不解析了
                        break;
                    case 'w:proofErr':
                    case 'w:noProof':
                        // 语法检查
                        break;
                    case 'w:del':
                    case 'w:moveTo':
                    case 'w:moveFrom':
                        // del 看起来主要是用于跟踪历史的，先不支持
                        break;
                    case 'w:fldSimple':
                        // 这个目前还不想支持
                        paragraph.fldSimples.push(FldSimple.FldSimple.fromXML(word, child));
                        break;
                    case 'm:oMathPara':
                    case 'm:oMath':
                        paragraph.addChild(OMath.OMath.fromXML(word, child));
                        break;
                    default:
                        console.warn('parse Paragraph: Unknown key', tagName, child);
                }
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return paragraph;
    };
    return Paragraph;
}());

exports.Paragraph = Paragraph;
