// office-viewer/lib/excel/render/formulaBar/FormulaBar.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var H = require('../../../util/H.js');
var Range = require('../../io/excel/util/Range.js');
var numberToLetters = require('../../io/excel/util/numberToLetters.js');
var CellData = require('../../types/worksheet/CellData.js');
var Input = require('../ui/Input.js');

/**
 * 公式栏
 */
var FormulaBar = /** @class */ (function () {
    function FormulaBar(dom, workbook, renderOptions) {
        var _this = this;
        this.dom = dom;
        this.workbook = workbook;
        if (!renderOptions.showFormulaBar) {
            return;
        }
        this.initDom();
        workbook.uiEvent.on('SWITCH_SHEET', function () {
            _this.sync();
        });
        workbook.uiEvent.on('CHANGE_SELECTION', function () {
            _this.sync();
        });
    }
    FormulaBar.prototype.initDom = function () {
        var _this = this;
        var nameBox = H.H('div', {
            className: 'ov-excel-formula-bar__name-box',
            parent: this.dom
        });
        this.nameBox = nameBox;
        var textBox = H.H('div', {
            className: 'ov-excel-formula-bar__text-box',
            parent: this.dom
        });
        this.textBox = textBox;
        var textInput = new Input.Input({
            container: textBox,
            onChange: function (value) {
                _this.changeCellValue(value);
            },
            style: 'borderLess'
        });
        this.textInput = textInput;
    };
    FormulaBar.prototype.getActiveCell = function () {
        var currentSheet = this.workbook.getActiveSheet();
        var selection = currentSheet.getSelection();
        if (!selection) {
            return null;
        }
        return selection.activeCell;
    };
    /**
     * 同步 dom 节点
     */
    FormulaBar.prototype.sync = function () {
        var currentSheet = this.workbook.getActiveSheet();
        var activeCell = this.getActiveCell();
        if (!activeCell) {
            return;
        }
        if (Range.isSingleCell(activeCell) || currentSheet.isMergeCell(activeCell)) {
            var cellInfo = this.workbook
                .getActiveSheet()
                .getCellInfo(activeCell.startRow, activeCell.startCol);
            this.textInput.setValue(cellInfo.text);
            this.cellData = currentSheet.getCellData(activeCell.startRow, activeCell.startCol);
        }
        else {
            this.textInput.setValue('');
        }
        this.nameBox.innerText = "".concat(numberToLetters.numberToLetters(activeCell.startCol)).concat(activeCell.startRow + 1);
    };
    FormulaBar.prototype.changeCellValue = function (value) {
        var currentSheet = this.workbook.getActiveSheet();
        var activeCell = this.getActiveCell();
        if (!activeCell) {
            return;
        }
        if (Range.isSingleCell(activeCell) || currentSheet.isMergeCell(activeCell)) {
            currentSheet.updateCellValue(activeCell.startRow, activeCell.startCol, CellData.updateValue(value, this.cellData));
            this.workbook.uiEvent.emit('UPDATE_RANGE', currentSheet.getIndex(), activeCell);
        }
    };
    return FormulaBar;
}());

exports.FormulaBar = FormulaBar;
