// amis-core/esm/store/app.js
/**
 * amis-core v6.13.0
 * Copyright 2018-2025 fex
 */

import { __assign } from 'tslib';
import { types } from 'mobx-state-tree';
import { mapTree, findTree, isVisible, guid, replaceUrlParams } from '../utils/helper.js';
import { ServiceStore } from './service.js';
import 'amis-formula';
import 'moment';
import { createObject } from '../utils/object.js';
import '../utils/memoryParse.js';
import '@rc-component/mini-decimal';
import 'lodash/isPlainObject';
import '../utils/filter.js';
import '../utils/api.js';
import 'file-saver';
import '../utils/browser.js';
import '../utils/ColorScale.js';
import 'react';
import 'lodash/chunk';
import '../utils/DataSchema.js';
import '../utils/DataScope.js';
import '../utils/debug.js';
import 'react-dom';
import '../utils/errors.js';
import { filter } from '../utils/tpl.js';
import 'lodash/isObject';
import 'lodash/isString';
import 'lodash/isBoolean';
import '../utils/image.js';
import 'lodash/isEqual';
import '../actions/Action.js';
import 'lodash/debounce';
import '../utils/resize-sensor.js';
import 'react-overlays/useRootClose';
import '../utils/SimpleMap.js';
import 'lodash/mapValues';
import 'lodash/camelCase';
import 'lodash/cloneDeep';
import 'lodash/map';
import 'lodash/isEmpty';
import 'lodash/kebabCase';
import 'uncontrollable';
import 'hoist-non-react-statics';
import '../utils/validations.js';
import '../utils/resolveCondition.js';
import 'mobx';
import '../utils/Animation.js';

var AppStore = ServiceStore.named('AppStore')
    .props({
    pages: types.frozen(),
    activePage: types.frozen(),
    folded: false,
    offScreen: false
})
    .views(function (self) { return ({
    get navigations() {
        if (Array.isArray(self.pages)) {
            return mapTree(self.pages, function (item) {
                var visible = isVisible(item, self.data);
                if (visible !== false &&
                    item.path &&
                    !~item.path.indexOf('http') &&
                    ~item.path.indexOf(':')) {
                    visible = false;
                }
                return {
                    label: item.label,
                    icon: item.icon,
                    path: item.path,
                    children: item.children,
                    className: item.className,
                    visible: visible,
                    badge: typeof item.badge === 'string'
                        ? filter(item.badge, self.data)
                        : item.badge,
                    badgeClassName: filter(item.badgeClassName, self.data)
                };
            });
        }
        return [
            {
                label: self.__('App.navigation'),
                children: []
            }
        ];
    },
    get bcn() {
        var _a;
        return ((_a = self.activePage) === null || _a === void 0 ? void 0 : _a.bcn) || [];
    },
    get pageData() {
        var _a;
        return createObject(self.data, {
            params: ((_a = self.activePage) === null || _a === void 0 ? void 0 : _a.params) || {}
        });
    }
}); })
    .actions(function (self) { return ({
    toggleFolded: function () {
        self.folded = !self.folded;
    },
    toggleOffScreen: function () {
        self.offScreen = !self.offScreen;
    },
    updateOffScreen: function (value) {
        self.offScreen = value;
    },
    setPages: function (pages) {
        if (pages && !Array.isArray(pages)) {
            pages = [pages];
        }
        else if (!Array.isArray(pages)) {
            return;
        }
        pages = mapTree(pages, function (item, index, level, paths) {
            var path = item.link || item.url;
            if (item.schema || item.schemaApi) {
                path =
                    item.url ||
                        "/".concat(paths
                            .map(function (item) { return item.index; })
                            .concat(index)
                            .map(function (index) { return "page-".concat(index + 1); })
                            .join('/'));
                if (path && path[0] !== '/') {
                    var parentPath = '/';
                    var index_1 = paths.length;
                    while (index_1 > 0) {
                        var item_1 = paths[index_1 - 1];
                        if (item_1 === null || item_1 === void 0 ? void 0 : item_1.path) {
                            parentPath = item_1.path + '/';
                            break;
                        }
                        index_1--;
                    }
                    path = parentPath + path;
                }
            }
            return __assign(__assign({}, item), { index: index, id: item.id || guid(), label: item.label, icon: item.icon, path: path });
        });
        self.pages = pages;
    },
    rewrite: function (to, env) {
        var page = findTree(self.pages, function (item) {
            if (item.path === to) {
                return true;
            }
            return false;
        });
        if (page) {
            this.setActivePage(page, env);
        }
    },
    setActivePage: function (page, env, params) {
        var _a;
        // 同一个页面直接返回。
        if (((_a = self.activePage) === null || _a === void 0 ? void 0 : _a.id) === page.id) {
            if (params !== undefined) {
                self.activePage = __assign(__assign({}, self.activePage), { params: params });
            }
            return;
        }
        var bcn = [];
        findTree(self.pages, function (item, index, level, paths) {
            if (item.id === page.id) {
                bcn = paths
                    .filter(function (item) { return item.path && item.label; })
                    .map(function (item) { return (__assign(__assign({}, item), { path: replaceUrlParams(item.path, params) })); });
                if (env.showFullBreadcrumbPath) {
                    bcn = paths.filter(function (item) { return item.label; });
                }
                bcn.push(__assign(__assign({}, item), { path: '' }));
                self.__;
                if (env.showBreadcrumbHomePath && bcn[0].path !== '/') {
                    bcn.unshift({
                        label: self.__('App.home'),
                        path: '/'
                    });
                }
                return true;
            }
            return false;
        });
        self.activePage = __assign(__assign({}, page), { params: params || {}, bcn: bcn });
        if (page.label) {
            document.title = page.label;
        }
        if (page.schema) {
            self.schema = page.schema;
            self.schemaKey = '' + Date.now();
        }
        else if (page.schemaApi) {
            self.schema = null;
            self.fetchSchema(page.schemaApi, self.activePage, { method: 'get' });
        }
        else if (page.redirect) {
            env.jumpTo(page.redirect, undefined, self.data);
            return;
        }
        else if (page.rewrite) {
            this.rewrite(page.rewrite, env);
        }
        else {
            self.schema = null;
            self.schemaKey = '';
        }
    },
    updateActivePage: function (env) {
        if (!Array.isArray(self.pages)) {
            return;
        }
        var matched;
        var page = findTree(self.pages, function (item) {
            if (item.path) {
                matched = env.isCurrentUrl(item.path, item);
                if (matched) {
                    return true;
                }
            }
            return false;
        });
        if (page) {
            this.setActivePage(page, env, typeof matched === 'object' ? matched.params : undefined);
        }
        else {
            var page_1 = findTree(self.pages, function (item) { return item.isDefaultPage; });
            if (page_1) {
                this.setActivePage(page_1, env);
            }
            else {
                self.activePage = null;
            }
        }
    }
}); });

export { AppStore };
