// amis-core/esm/store/status.js
/**
 * amis-core v6.13.0
 * Copyright 2018-2025 fex
 */

import { __assign } from 'tslib';
import { types } from 'mobx-state-tree';

var StatusStore = types
    .model('StatusStore', {
    visibleState: types.optional(types.frozen(), {}),
    disableState: types.optional(types.frozen(), {}),
    staticState: types.optional(types.frozen(), {})
})
    .actions(function (self) { return ({
    setVisible: function (key, value) {
        var _a;
        var state = __assign(__assign({}, self.visibleState), (_a = {}, _a[key] = value, _a));
        self.visibleState = state;
    },
    setDisable: function (key, value) {
        var _a;
        var state = __assign(__assign({}, self.disableState), (_a = {}, _a[key] = value, _a));
        self.disableState = state;
    },
    setStatic: function (key, value) {
        var _a;
        var state = __assign(__assign({}, self.staticState), (_a = {}, _a[key] = value, _a));
        self.staticState = state;
    },
    resetAll: function () {
        self.visibleState = {};
        self.disableState = {};
        self.staticState = {};
    }
}); });

export { StatusStore };
