// office-viewer/esm/excel/render/autoFilter/filterColumn/FiltersUI.js
import { H } from '../../../../util/H.js';
import { CheckBoxList } from '../../ui/CheckBoxList.js';
import { getFilterValues } from './getFilterValues.js';
import { setFilterValues } from './setFilterValues.js';

/**
 * 按文本分类进行过滤
 */
var FiltersUI = /** @class */ (function () {
    function FiltersUI(autoFilterIcon, container, sheet, autoFilter, colIndex, headerRowCount, texts) {
        var filtersContainer = H('div', {
            className: 'ov-excel-auto-filter__menu-filter-filters',
            parent: container
        });
        this.filtersContainer = filtersContainer;
        var filterValues = getFilterValues(autoFilter, colIndex);
        var selectAll = filterValues.size === 0;
        var options = texts.map(function (text) {
            return {
                text: text,
                value: text,
                checked: selectAll || filterValues.has(text)
            };
        });
        var selectAllText = sheet.translate('checkBox.selectAll');
        var searchPlaceholder = sheet.translate('checkBox.search');
        new CheckBoxList(filtersContainer, selectAllText, searchPlaceholder, options, function (checked, option, options) {
            setFilterValues(autoFilter, colIndex, options);
            sheet.applyAutoFilter(autoFilter, headerRowCount);
            sheet.getWorkbook().uiEvent.emit('APPLY_AUTO_FILTER', sheet.getIndex());
            autoFilterIcon.syncFilterIcon();
        });
    }
    return FiltersUI;
}());

export { FiltersUI };
