// office-viewer/lib/excel/sheet/table/buildTableStyle.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');

/**
 * 构建演示表，这个主要原因是样式应用有顺序要求，所以先构建为对象格式
 */
function buildTableStyle(tableStyle) {
    var e_1, _a;
    var _b, _c, _d;
    var wholeTable;
    var headerRow;
    var firstRowStripe;
    var firstColumnStripe;
    var totalRow;
    var firstColumn;
    var lastColumn;
    try {
        for (var _e = tslib.__values(((_c = (_b = tableStyle.tableStyles.tableStyle) === null || _b === void 0 ? void 0 : _b[0]) === null || _c === void 0 ? void 0 : _c.tableStyleElement) || []), _f = _e.next(); !_f.done; _f = _e.next()) {
            var tableStyleElement = _f.value;
            var type = tableStyleElement.type;
            var dxfId = (_d = tableStyleElement.dxfId) !== null && _d !== void 0 ? _d : 1;
            // 文档里说是从 0 开始，但其实是从 1 开始
            dxfId = dxfId - 1;
            switch (type) {
                case 'wholeTable':
                    wholeTable = dxfId;
                    break;
                case 'headerRow':
                    headerRow = dxfId;
                    break;
                case 'firstRowStripe':
                    firstRowStripe = dxfId;
                    break;
                case 'firstColumnStripe':
                    firstColumnStripe = dxfId;
                    break;
                case 'totalRow':
                    totalRow = dxfId;
                    break;
                case 'firstColumn':
                    firstColumn = dxfId;
                    break;
                case 'lastColumn':
                    lastColumn = dxfId;
                    break;
                default:
                    break;
            }
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (_f && !_f.done && (_a = _e.return)) _a.call(_e);
        }
        finally { if (e_1) throw e_1.error; }
    }
    return {
        wholeTable: wholeTable,
        headerRow: headerRow,
        firstRowStripe: firstRowStripe,
        firstColumnStripe: firstColumnStripe,
        totalRow: totalRow,
        firstColumn: firstColumn,
        lastColumn: lastColumn
    };
}

exports.buildTableStyle = buildTableStyle;
