// amis-core/lib/store/app.js
/**
 * amis-core v6.13.0
 * Copyright 2018-2025 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var mobxStateTree = require('mobx-state-tree');
var helper = require('../utils/helper.js');
var service = require('./service.js');
require('amis-formula');
require('moment');
var object = require('../utils/object.js');
require('../utils/memoryParse.js');
require('@rc-component/mini-decimal');
require('lodash/isPlainObject');
require('../utils/filter.js');
require('../utils/api.js');
require('file-saver');
require('../utils/browser.js');
require('../utils/ColorScale.js');
require('../utils/columnsSplit.js');
require('../utils/DataSchema.js');
require('../utils/DataScope.js');
require('../utils/debug.js');
require('../utils/dom.js');
require('../utils/errors.js');
var tpl = require('../utils/tpl.js');
require('lodash/isObject');
require('lodash/isString');
require('lodash/isBoolean');
require('../utils/highlight.js');
require('../utils/icon.js');
require('../utils/image.js');
require('lodash/isEqual');
require('../actions/Action.js');
require('lodash/debounce');
require('../utils/resize-sensor.js');
require('react');
require('react-overlays/useRootClose');
require('react-dom');
require('../utils/SimpleMap.js');
require('lodash/mapValues');
require('lodash/camelCase');
require('lodash/cloneDeep');
require('lodash/map');
require('lodash/isEmpty');
require('lodash/kebabCase');
require('uncontrollable');
require('hoist-non-react-statics');
require('../utils/validations.js');
require('../utils/resolveCondition.js');
require('mobx');
require('../utils/Animation.js');

var AppStore = service.ServiceStore.named('AppStore')
    .props({
    pages: mobxStateTree.types.frozen(),
    activePage: mobxStateTree.types.frozen(),
    folded: false,
    offScreen: false
})
    .views(function (self) { return ({
    get navigations() {
        if (Array.isArray(self.pages)) {
            return helper.mapTree(self.pages, function (item) {
                var visible = helper.isVisible(item, self.data);
                if (visible !== false &&
                    item.path &&
                    !~item.path.indexOf('http') &&
                    ~item.path.indexOf(':')) {
                    visible = false;
                }
                return {
                    label: item.label,
                    icon: item.icon,
                    path: item.path,
                    children: item.children,
                    className: item.className,
                    visible: visible,
                    badge: typeof item.badge === 'string'
                        ? tpl.filter(item.badge, self.data)
                        : item.badge,
                    badgeClassName: tpl.filter(item.badgeClassName, self.data)
                };
            });
        }
        return [
            {
                label: self.__('App.navigation'),
                children: []
            }
        ];
    },
    get bcn() {
        var _a;
        return ((_a = self.activePage) === null || _a === void 0 ? void 0 : _a.bcn) || [];
    },
    get pageData() {
        var _a;
        return object.createObject(self.data, {
            params: ((_a = self.activePage) === null || _a === void 0 ? void 0 : _a.params) || {}
        });
    }
}); })
    .actions(function (self) { return ({
    toggleFolded: function () {
        self.folded = !self.folded;
    },
    toggleOffScreen: function () {
        self.offScreen = !self.offScreen;
    },
    updateOffScreen: function (value) {
        self.offScreen = value;
    },
    setPages: function (pages) {
        if (pages && !Array.isArray(pages)) {
            pages = [pages];
        }
        else if (!Array.isArray(pages)) {
            return;
        }
        pages = helper.mapTree(pages, function (item, index, level, paths) {
            var path = item.link || item.url;
            if (item.schema || item.schemaApi) {
                path =
                    item.url ||
                        "/".concat(paths
                            .map(function (item) { return item.index; })
                            .concat(index)
                            .map(function (index) { return "page-".concat(index + 1); })
                            .join('/'));
                if (path && path[0] !== '/') {
                    var parentPath = '/';
                    var index_1 = paths.length;
                    while (index_1 > 0) {
                        var item_1 = paths[index_1 - 1];
                        if (item_1 === null || item_1 === void 0 ? void 0 : item_1.path) {
                            parentPath = item_1.path + '/';
                            break;
                        }
                        index_1--;
                    }
                    path = parentPath + path;
                }
            }
            return tslib.__assign(tslib.__assign({}, item), { index: index, id: item.id || helper.guid(), label: item.label, icon: item.icon, path: path });
        });
        self.pages = pages;
    },
    rewrite: function (to, env) {
        var page = helper.findTree(self.pages, function (item) {
            if (item.path === to) {
                return true;
            }
            return false;
        });
        if (page) {
            this.setActivePage(page, env);
        }
    },
    setActivePage: function (page, env, params) {
        var _a;
        // 同一个页面直接返回。
        if (((_a = self.activePage) === null || _a === void 0 ? void 0 : _a.id) === page.id) {
            if (params !== undefined) {
                self.activePage = tslib.__assign(tslib.__assign({}, self.activePage), { params: params });
            }
            return;
        }
        var bcn = [];
        helper.findTree(self.pages, function (item, index, level, paths) {
            if (item.id === page.id) {
                bcn = paths
                    .filter(function (item) { return item.path && item.label; })
                    .map(function (item) { return (tslib.__assign(tslib.__assign({}, item), { path: helper.replaceUrlParams(item.path, params) })); });
                if (env.showFullBreadcrumbPath) {
                    bcn = paths.filter(function (item) { return item.label; });
                }
                bcn.push(tslib.__assign(tslib.__assign({}, item), { path: '' }));
                self.__;
                if (env.showBreadcrumbHomePath && bcn[0].path !== '/') {
                    bcn.unshift({
                        label: self.__('App.home'),
                        path: '/'
                    });
                }
                return true;
            }
            return false;
        });
        self.activePage = tslib.__assign(tslib.__assign({}, page), { params: params || {}, bcn: bcn });
        if (page.label) {
            document.title = page.label;
        }
        if (page.schema) {
            self.schema = page.schema;
            self.schemaKey = '' + Date.now();
        }
        else if (page.schemaApi) {
            self.schema = null;
            self.fetchSchema(page.schemaApi, self.activePage, { method: 'get' });
        }
        else if (page.redirect) {
            env.jumpTo(page.redirect, undefined, self.data);
            return;
        }
        else if (page.rewrite) {
            this.rewrite(page.rewrite, env);
        }
        else {
            self.schema = null;
            self.schemaKey = '';
        }
    },
    updateActivePage: function (env) {
        if (!Array.isArray(self.pages)) {
            return;
        }
        var matched;
        var page = helper.findTree(self.pages, function (item) {
            if (item.path) {
                matched = env.isCurrentUrl(item.path, item);
                if (matched) {
                    return true;
                }
            }
            return false;
        });
        if (page) {
            this.setActivePage(page, env, typeof matched === 'object' ? matched.params : undefined);
        }
        else {
            var page_1 = helper.findTree(self.pages, function (item) { return item.isDefaultPage; });
            if (page_1) {
                this.setActivePage(page_1, env);
            }
            else {
                self.activePage = null;
            }
        }
    }
}); });

exports.AppStore = AppStore;
