// amis-core/lib/store/iRenderer.js
/**
 * amis-core v6.13.0
 * Copyright 2018-2025 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var mobxStateTree = require('mobx-state-tree');
require('../utils/helper.js');
require('amis-formula');
require('moment');
var object = require('../utils/object.js');
require('../utils/memoryParse.js');
require('@rc-component/mini-decimal');
var getVariable = require('../utils/getVariable.js');
var dataMapping = require('../utils/dataMapping.js');
require('../utils/filter.js');
var SimpleMap = require('../utils/SimpleMap.js');
var node = require('./node.js');
require('../utils/api.js');
require('file-saver');
require('../utils/browser.js');
require('../utils/ColorScale.js');
require('../utils/columnsSplit.js');
var concatData = require('../utils/concatData.js');
require('../utils/DataSchema.js');
require('../utils/DataScope.js');
require('../utils/debug.js');
require('../utils/dom.js');
require('../utils/errors.js');
require('../utils/tpl.js');
require('lodash/isPlainObject');
require('lodash/isObject');
require('lodash/isString');
require('lodash/isBoolean');
require('../utils/highlight.js');
require('../utils/icon.js');
require('../utils/image.js');
require('lodash/isEqual');
require('../actions/Action.js');
require('lodash/debounce');
require('../utils/resize-sensor.js');
require('react');
require('react-overlays/useRootClose');
require('react-dom');
require('lodash/mapValues');
require('lodash/camelCase');
require('lodash/cloneDeep');
require('lodash/map');
require('lodash/isEmpty');
require('lodash/kebabCase');
require('uncontrollable');
require('hoist-non-react-statics');
require('../utils/validations.js');
require('../utils/resolveCondition.js');
require('mobx');
require('../utils/Animation.js');
var findLastIndex = require('lodash/findLastIndex');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var findLastIndex__default = /*#__PURE__*/_interopDefaultLegacy(findLastIndex);

var iRendererStore = node.StoreNode.named('iRendererStore')
    .props({
    hasRemoteData: mobxStateTree.types.optional(mobxStateTree.types.boolean, false),
    data: mobxStateTree.types.optional(mobxStateTree.types.frozen(), {}),
    initedAt: 0,
    updatedAt: 0,
    pristine: mobxStateTree.types.optional(mobxStateTree.types.frozen(), {}),
    pristineRaw: mobxStateTree.types.optional(mobxStateTree.types.frozen(), {}),
    upStreamData: mobxStateTree.types.optional(mobxStateTree.types.frozen(), {}),
    action: mobxStateTree.types.optional(mobxStateTree.types.frozen(), undefined),
    dialogSchema: mobxStateTree.types.frozen(),
    dialogOpen: false,
    dialogData: mobxStateTree.types.optional(mobxStateTree.types.frozen(), undefined),
    drawerSchema: mobxStateTree.types.frozen(),
    drawerOpen: false,
    drawerData: mobxStateTree.types.optional(mobxStateTree.types.frozen(), undefined)
})
    .views(function (self) { return ({
    getValueByName: function (name, canAccessSuper) {
        if (canAccessSuper === void 0) { canAccessSuper = true; }
        return getVariable.getVariable(self.data, name, canAccessSuper);
    },
    getPristineValueByName: function (name) {
        return getVariable.getVariable(self.pristine, name, false);
    },
    get pristineDiff() {
        var data = {};
        Object.keys(self.pristine).forEach(function (key) {
            if (self.pristine[key] !== self.pristineRaw[key]) {
                data[key] = self.pristine[key];
            }
        });
        return data;
    }
}); })
    .actions(function (self) {
    var dialogCallbacks = new SimpleMap.SimpleMap();
    var dialogScoped = null;
    var drawerScoped = null;
    var top = null;
    return {
        setTopStore: function (value) {
            top = value;
        },
        initData: function (data, skipSetPristine, changeReason) {
            if (data === void 0) { data = {}; }
            if (skipSetPristine === void 0) { skipSetPristine = false; }
            self.initedAt = Date.now();
            if (self.data.__tag) {
                data = object.injectObjectChain(data, self.data.__tag);
            }
            if (!skipSetPristine) {
                self.pristine = data;
                self.pristineRaw = data;
            }
            changeReason &&
                Object.isExtensible(data) &&
                !data.__changeReason &&
                Object.defineProperty(data, '__changeReason', {
                    value: changeReason,
                    enumerable: false,
                    configurable: false,
                    writable: false
                });
            self.data = data;
            self.upStreamData = data;
        },
        // 临时更新全局变量
        temporaryUpdateGlobalVars: function (globalVar) {
            var chain = object.extractObjectChain(self.data).filter(function (item) { return !item.hasOwnProperty('__isTempGlobalLayer'); });
            var idx = findLastIndex__default["default"](chain, function (item) {
                return item.hasOwnProperty('global') || item.hasOwnProperty('globalState');
            });
            if (idx !== -1) {
                chain.splice(idx + 1, 0, tslib.__assign(tslib.__assign({}, globalVar), { __isTempGlobalLayer: true }));
            }
            self.data = object.createObjectFromChain(chain);
        },
        // 撤销临时更新全局变量
        unDoTemporaryUpdateGlobalVars: function () {
            var chain = object.extractObjectChain(self.data).filter(function (item) { return !item.hasOwnProperty('__isTempGlobalLayer'); });
            self.data = object.createObjectFromChain(chain);
        },
        reset: function () {
            self.data = self.pristine;
        },
        updateData: function (data, tag, replace, concatFields, changeReason) {
            if (data === void 0) { data = {}; }
            if (concatFields) {
                data = concatData.concatData(data, self.data, concatFields);
            }
            var prev = self.data;
            var newData;
            if (tag) {
                var proto = object.createObject(self.data.__super || null, tslib.__assign(tslib.__assign({}, tag), { __tag: tag }));
                newData = object.createObject(proto, tslib.__assign(tslib.__assign({}, (replace ? {} : self.data)), data));
            }
            else {
                newData = object.extendObject(self.data, data, !replace);
            }
            Object.defineProperty(newData, '__prev', {
                value: tslib.__assign({}, prev),
                enumerable: false,
                configurable: false,
                writable: false
            });
            changeReason &&
                Object.isExtensible(newData) &&
                !newData.__changeReason &&
                Object.defineProperty(newData, '__changeReason', {
                    value: changeReason,
                    enumerable: false,
                    configurable: false,
                    writable: false
                });
            self.data = newData;
        },
        changeValue: function (name, value, changePristine, force, otherModifier, changeReason) {
            if (!name) {
                return;
            }
            var origin = getVariable.getVariable(self.data, name, false);
            if (value === origin && !force) {
                return;
            }
            var prev = self.data;
            var data = object.cloneObject(self.data);
            if (prev.hasOwnProperty('__prev')) {
                // 基于之前的 __prev 改
                var prevData = object.cloneObject(prev.__prev);
                object.setVariable(prevData, name, origin);
                Object.defineProperty(data, '__prev', {
                    value: prevData,
                    enumerable: false,
                    configurable: false,
                    writable: false
                });
            }
            else {
                Object.defineProperty(data, '__prev', {
                    value: tslib.__assign({}, prev),
                    enumerable: false,
                    configurable: false,
                    writable: false
                });
            }
            if (value === undefined) {
                object.deleteVariable(data, name);
            }
            else {
                object.setVariable(data, name, value);
            }
            otherModifier === null || otherModifier === void 0 ? void 0 : otherModifier(data);
            if (changePristine) {
                var pristine = object.cloneObject(self.pristine);
                object.setVariable(pristine, name, value);
                otherModifier === null || otherModifier === void 0 ? void 0 : otherModifier(pristine);
                self.pristine = pristine;
            }
            if (!data.__pristine) {
                Object.defineProperty(data, '__pristine', {
                    value: self.pristine,
                    enumerable: false,
                    configurable: false,
                    writable: false
                });
            }
            changeReason &&
                Object.isExtensible(data) &&
                !data.__changeReason &&
                Object.defineProperty(data, '__changeReason', {
                    value: changeReason,
                    enumerable: false,
                    configurable: false,
                    writable: false
                });
            self.data = data;
        },
        setCurrentAction: function (action, resolveDefinitions) {
            // 处理 $ref
            resolveDefinitions &&
                ['dialog', 'drawer'].forEach(function (key) {
                    var _a;
                    var _b;
                    if ((_b = action[key]) === null || _b === void 0 ? void 0 : _b.$ref) {
                        action = tslib.__assign(tslib.__assign({}, action), (_a = {}, _a[key] = tslib.__assign(tslib.__assign({}, resolveDefinitions(action[key].$ref)), action[key]), _a));
                    }
                });
            self.action = action;
        },
        openDialog: function (ctx, additonal, callback, scoped) {
            var _a, _b;
            var chain = object.extractObjectChain(ctx);
            chain.length === 1 && chain.unshift(self.data);
            if (additonal) {
                chain.splice(chain.length - 1, 0, additonal);
            }
            var data = object.createObjectFromChain(chain);
            var mappingData = (_a = self.action.data) !== null && _a !== void 0 ? _a : (_b = self.action.dialog) === null || _b === void 0 ? void 0 : _b.data;
            if (mappingData) {
                self.dialogData = object.createObjectFromChain([
                    top === null || top === void 0 ? void 0 : top.context,
                    dataMapping.dataMapping(mappingData, data)
                ]);
                var clonedAction = tslib.__assign(tslib.__assign({}, self.action), { dialog: tslib.__assign({}, self.action.dialog) });
                delete clonedAction.dialog.data;
                self.action = clonedAction;
            }
            else {
                self.dialogData = data;
            }
            self.dialogSchema = self.action.dialog;
            self.dialogOpen = true;
            callback && dialogCallbacks.set(self.dialogData, callback);
            dialogScoped = scoped || null;
        },
        closeDialog: function (confirmed, data) {
            var callback = dialogCallbacks.get(self.dialogData);
            // 不要过早的清空，否则内部组件提前销毁，会出现 store 异常读取问题
            // self.dialogSchema = null;
            self.dialogOpen = false;
            dialogScoped = null;
            if (callback) {
                dialogCallbacks.delete(self.dialogData);
                setTimeout(function () { return callback(confirmed, data); }, 200);
            }
        },
        openDrawer: function (ctx, additonal, callback, scoped) {
            var _a;
            var chain = object.extractObjectChain(ctx);
            chain.length === 1 && chain.unshift(self.data);
            if (additonal) {
                chain.splice(chain.length - 1, 0, additonal);
            }
            var data = object.createObjectFromChain(chain);
            var mappingData = (_a = self.action.data) !== null && _a !== void 0 ? _a : self.action.drawer.data;
            if (mappingData) {
                self.drawerData = object.createObjectFromChain([
                    top === null || top === void 0 ? void 0 : top.context,
                    dataMapping.dataMapping(mappingData, data)
                ]);
                var clonedAction = tslib.__assign(tslib.__assign({}, self.action), { drawer: tslib.__assign({}, self.action.drawer) });
                delete clonedAction.drawer.data;
                self.action = clonedAction;
            }
            else {
                self.drawerData = data;
            }
            self.drawerSchema = self.action.drawer;
            self.drawerOpen = true;
            if (callback) {
                dialogCallbacks.set(self.drawerData, callback);
            }
            drawerScoped = scoped || null;
        },
        closeDrawer: function (confirmed, data) {
            var callback = dialogCallbacks.get(self.drawerData);
            // 不要过早的清空，否则内部组件提前销毁，会出现 store 异常读取问题
            // self.drawerSchema = null;
            self.drawerOpen = false;
            drawerScoped = null;
            if (callback) {
                dialogCallbacks.delete(self.drawerData);
                setTimeout(function () { return callback(confirmed, data); }, 200);
            }
        },
        getDialogScoped: function () {
            return dialogScoped;
        },
        getDrawerScoped: function () {
            return drawerScoped;
        }
    };
});
// export type SIRendererStore = typeof iRendererStore.SnapshotType;

exports.iRendererStore = iRendererStore;
