// amis-core/lib/store/modal.js
/**
 * amis-core v6.13.0
 * Copyright 2018-2025 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var service = require('./service.js');
var mobxStateTree = require('mobx-state-tree');
require('../utils/helper.js');
var object = require('../utils/object.js');

var ModalStore = service.ServiceStore.named('ModalStore')
    .props({
    form: mobxStateTree.types.frozen(),
    entered: false,
    dragging: false,
    resizeCoord: 0,
    schema: mobxStateTree.types.frozen(),
    isFullscreen: false
})
    .views(function (self) {
    return {
        get formData() {
            return object.createObject(self.data, self.form);
        },
        get inDragging() {
            return self.dragging;
        }
    };
})
    .actions(function (self) {
    return {
        setEntered: function (value) {
            self.entered = value;
        },
        setDragging: function (value) {
            self.dragging = value;
        },
        setFormData: function (obj) {
            self.form = obj;
        },
        reset: function () {
            self.form = {};
            self.reInitData({}, true);
        },
        setResizeCoord: function (value) {
            self.resizeCoord = value;
        },
        setSchema: function (schema) {
            if (schema && schema.then) {
                schema.then(function (value) { return mobxStateTree.isAlive(self) && self.setSchema(value); });
                return;
            }
            self.schema = schema;
        },
        setFullScreen: function (isFullscreen) {
            if (isFullscreen === void 0) { isFullscreen = false; }
            self.isFullscreen = isFullscreen;
        }
    };
});

exports.ModalStore = ModalStore;
