// amis-core/lib/store/status.js
/**
 * amis-core v6.13.0
 * Copyright 2018-2025 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var mobxStateTree = require('mobx-state-tree');

var StatusStore = mobxStateTree.types
    .model('StatusStore', {
    visibleState: mobxStateTree.types.optional(mobxStateTree.types.frozen(), {}),
    disableState: mobxStateTree.types.optional(mobxStateTree.types.frozen(), {}),
    staticState: mobxStateTree.types.optional(mobxStateTree.types.frozen(), {})
})
    .actions(function (self) { return ({
    setVisible: function (key, value) {
        var _a;
        var state = tslib.__assign(tslib.__assign({}, self.visibleState), (_a = {}, _a[key] = value, _a));
        self.visibleState = state;
    },
    setDisable: function (key, value) {
        var _a;
        var state = tslib.__assign(tslib.__assign({}, self.disableState), (_a = {}, _a[key] = value, _a));
        self.disableState = state;
    },
    setStatic: function (key, value) {
        var _a;
        var state = tslib.__assign(tslib.__assign({}, self.staticState), (_a = {}, _a[key] = value, _a));
        self.staticState = state;
    },
    resetAll: function () {
        self.visibleState = {};
        self.disableState = {};
        self.staticState = {};
    }
}); });

exports.StatusStore = StatusStore;
