// office-viewer/lib/excel/render/grid/drawGridLines.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/**
 * 生成网格线
 */
function generateGridLines(currentSheet, viewRange, width, height) {
    var rows = viewRange.rows, rowSizes = viewRange.rowSizes, cols = viewRange.cols, colSizes = viewRange.colSizes;
    var lines = [];
    // 水平线
    for (var index = 0; index < rows.length; index++) {
        var currentRowOffset = rowSizes[index].offset;
        lines.push({
            x1: 0,
            y1: currentRowOffset,
            x2: width,
            y2: currentRowOffset
        });
    }
    // 垂直线
    for (var index = 0; index < cols.length; index++) {
        var currentColOffset = colSizes[index].offset;
        lines.push({
            x1: currentColOffset,
            y1: 0,
            x2: currentColOffset,
            y2: height
        });
    }
    return lines;
}
/**
 * 绘制网格线
 */
function drawGridLines(currentSheet, viewRange, canvas, height, width, gridLineOptions) {
    if (!currentSheet.showGridLines()) {
        return;
    }
    var lines = generateGridLines(currentSheet, viewRange, width, height);
    canvas.drawLines(lines, gridLineOptions.gridLineColor);
}

exports.drawGridLines = drawGridLines;
exports.generateGridLines = generateGridLines;
