// office-viewer/lib/excel/render/autoFilter/AutoFilterMenuUI.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var H = require('../../../util/H.js');
var FilterColumnUI = require('./filterColumn/FilterColumnUI.js');
var SortStateUI = require('./sortState/SortStateUI.js');

var AutoFilterMenuUI = /** @class */ (function () {
    function AutoFilterMenuUI(autoFilterIcon, container, sheet, autoFilter, rangeRef, colIndex, headerRowCount) {
        if (headerRowCount === void 0) { headerRowCount = 1; }
        var menuContainer = H.H('div', {
            className: 'ov-excel-auto-filter-menu',
            parent: container
        });
        this.menuContainer = menuContainer;
        new SortStateUI.SortStateUI(autoFilterIcon, menuContainer, sheet, autoFilter, rangeRef, colIndex, headerRowCount);
        new FilterColumnUI.FilterColumnUI(autoFilterIcon, menuContainer, sheet, autoFilter, rangeRef, colIndex, headerRowCount);
        this.autoFilter = autoFilter;
        this.colIndex = colIndex;
        this.sheet = sheet;
        menuContainer.addEventListener('mousedown', function (e) {
            e.stopPropagation();
        });
        menuContainer.addEventListener('wheel', function (e) {
            e.stopPropagation();
        });
    }
    AutoFilterMenuUI.prototype.hide = function () {
        this.menuContainer.style.display = 'none';
    };
    AutoFilterMenuUI.prototype.show = function () {
        this.menuContainer.style.display = 'flex';
    };
    return AutoFilterMenuUI;
}());

exports.AutoFilterMenuUI = AutoFilterMenuUI;
