// office-viewer/esm/excel/render/drawing/getRectFromAnchorPoint.js
import { emuToPx } from '../../../util/emuToPx.js';
import { findPositionInViewRange } from './findPositionInViewRange.js';

function getRectFromOneAnchorPoint(currentSheet, cell, viewRange) {
    var _a, _b, _c, _d;
    var cellRow = ((_a = cell.row) === null || _a === void 0 ? void 0 : _a[0]) || 0;
    var cellRowOffset = emuToPx((_b = cell.rowOff) === null || _b === void 0 ? void 0 : _b[0]);
    var cellCol = ((_c = cell.col) === null || _c === void 0 ? void 0 : _c[0]) || 0;
    var cellColOffset = emuToPx((_d = cell.colOff) === null || _d === void 0 ? void 0 : _d[0]);
    var cellPosition = findPositionInViewRange(currentSheet, cellRow, cellCol, viewRange);
    return {
        x: cellPosition.x,
        y: cellPosition.y,
        width: cellColOffset,
        height: cellRowOffset
    };
}
/**
 * 基于 Anchor Point 计算位置信息
 */
function getRectFromTwoAnchorPoint(currentSheet, from, to, viewRange) {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    var fromRow = ((_a = from.row) === null || _a === void 0 ? void 0 : _a[0]) || 0;
    var fromRowOffset = emuToPx((_b = from.rowOff) === null || _b === void 0 ? void 0 : _b[0]);
    var fromCol = ((_c = from.col) === null || _c === void 0 ? void 0 : _c[0]) || 0;
    var fromColOffset = emuToPx((_d = from.colOff) === null || _d === void 0 ? void 0 : _d[0]);
    var toRow = ((_e = to.row) === null || _e === void 0 ? void 0 : _e[0]) || 0;
    var toRowOffset = emuToPx((_f = to.rowOff) === null || _f === void 0 ? void 0 : _f[0]);
    var toCol = ((_g = to.col) === null || _g === void 0 ? void 0 : _g[0]) || 0;
    var toColOffset = emuToPx((_h = to.colOff) === null || _h === void 0 ? void 0 : _h[0]);
    var fromPosition = findPositionInViewRange(currentSheet, fromRow, fromCol, viewRange);
    var toPosition = findPositionInViewRange(currentSheet, toRow, toCol, viewRange);
    var x = fromPosition.x + fromColOffset;
    var y = fromPosition.y + fromRowOffset;
    return {
        x: x,
        y: y,
        width: toPosition.x + toColOffset - x,
        height: toPosition.y + toRowOffset - y
    };
}

export { getRectFromOneAnchorPoint, getRectFromTwoAnchorPoint };
