// amis-core/esm/components/LazyComponent.js
/**
 * amis-core v6.13.0
 * Copyright 2018-2025 fex
 */

import { __extends, __rest, __assign } from 'tslib';
import React from 'react';
import { InView } from 'react-intersection-observer';
import { themeable } from '../theme.js';

/**
 * @file LazyComponent
 * @description
 * @author fex
 */
var LazyComponent = /** @class */ (function (_super) {
    __extends(LazyComponent, _super);
    function LazyComponent(props) {
        var _this = this;
        var _a;
        _this = _super.call(this, props) || this;
        _this.mounted = false;
        _this.handleVisibleChange = _this.handleVisibleChange.bind(_this);
        _this.mounted = true;
        _this.state = {
            visible: (_a = props.defaultVisible) !== null && _a !== void 0 ? _a : false,
            component: props.component
        };
        return _this;
    }
    LazyComponent.prototype.componentDidMount = function () {
        // jest 里面有点异常，先手动让它总是可见
        if (typeof jest !== 'undefined' || this.state.visible) {
            this.handleVisibleChange(true);
        }
    };
    LazyComponent.prototype.componentWillUnmount = function () {
        this.mounted = false;
    };
    LazyComponent.prototype.handleVisibleChange = function (visible, entry) {
        var _this = this;
        this.setState({
            visible: visible
        });
        if (!visible || this.state.component || !this.props.getComponent) {
            return;
        }
        this.props
            .getComponent()
            .then(function (component) {
            return _this.mounted &&
                typeof component === 'function' &&
                _this.setState({
                    component: component
                });
        })
            .catch(function (reason) {
            return _this.mounted &&
                _this.setState({
                    component: function () { return (React.createElement("div", { className: "alert alert-danger" }, String(reason))); }
                });
        });
    };
    LazyComponent.prototype.render = function () {
        var _a = this.props, placeholder = _a.placeholder, unMountOnHidden = _a.unMountOnHidden, childProps = _a.childProps, partialVisibility = _a.partialVisibility, children = _a.children, className = _a.className, rest = __rest(_a, ["placeholder", "unMountOnHidden", "childProps", "partialVisibility", "children", "className"]);
        var cx = this.props.classnames;
        var _b = this.state, visible = _b.visible, Component = _b.component;
        // 需要监听从可见到不可见。
        if (unMountOnHidden) {
            return (React.createElement(InView, { onChange: this.handleVisibleChange, threshold: partialVisibility ? 0 : 1 }, function (_a) {
                var ref = _a.ref;
                return (React.createElement("div", { ref: ref, className: "visibility-sensor ".concat(visible ? 'in' : '') }, Component && visible ? (React.createElement(Component, __assign({}, rest, childProps))) : children && visible ? (children) : (placeholder)));
            }));
        }
        if (!visible) {
            return (React.createElement(InView, { onChange: this.handleVisibleChange, threshold: partialVisibility ? 0 : 1 }, function (_a) {
                var ref = _a.ref;
                return (React.createElement("div", { ref: ref, className: "visibility-sensor" }, placeholder));
            }));
        }
        else if (Component) {
            // 只监听不可见到可见，一旦可见了，就销毁检查。
            return React.createElement(Component, __assign({}, rest, childProps));
        }
        else if (children) {
            return children;
        }
        return React.createElement("div", { className: cx('LazyComponent', className) }, placeholder);
    };
    LazyComponent.defaultProps = {
        placeholder: React.createElement("span", null, "Loading..."),
        unMountOnHidden: false,
        partialVisibility: true
    };
    return LazyComponent;
}(React.Component));
var themedLazyComponent = themeable(LazyComponent);
themedLazyComponent.defaultProps = LazyComponent.defaultProps;

export { LazyComponent, themedLazyComponent as default };
