// amis-core/esm/components/Overlay.js
/**
 * amis-core v6.13.0
 * Copyright 2018-2025 fex
 */

import { __extends, __assign, __rest, __decorate, __metadata } from 'tslib';
import Portal from 'react-overlays/Portal';
import cx from 'classnames';
import ReactDOM, { findDOMNode } from 'react-dom';
import React, { cloneElement } from 'react';
import 'amis-formula';
import 'moment';
import 'lodash/isPlainObject';
import '../utils/memoryParse.js';
import '@rc-component/mini-decimal';
import '../utils/filter.js';
import '../utils/api.js';
import 'file-saver';
import '../utils/browser.js';
import '../utils/ColorScale.js';
import 'lodash/chunk';
import '../utils/DataSchema.js';
import '../utils/DataScope.js';
import '../utils/debug.js';
import { getContainer, ownerDocument, calculatePosition } from '../utils/dom.js';
import '../utils/errors.js';
import '../utils/tpl.js';
import { uuid, noop, getScrollParent, autobind } from '../utils/helper.js';
import 'lodash/isObject';
import 'lodash/isString';
import 'lodash/isBoolean';
import '../utils/image.js';
import 'lodash/isEqual';
import '../actions/Action.js';
import 'lodash/debounce';
import { resizeSensor } from '../utils/resize-sensor.js';
import { RootClose } from '../utils/RootClose.js';
import '../utils/SimpleMap.js';
import 'lodash/mapValues';
import 'lodash/camelCase';
import 'lodash/cloneDeep';
import 'lodash/map';
import 'lodash/isEmpty';
import 'lodash/kebabCase';
import 'uncontrollable';
import 'hoist-non-react-statics';
import '../utils/validations.js';
import '../utils/resolveCondition.js';
import 'mobx';
import '../utils/Animation.js';
import { EnvContext } from '../env.js';

/**
 * @file Overlay
 * @description
 * @author fex
 */
var SubPopoverDisplayedID = 'data-sub-popover-displayed';
function onScroll(elem, callback) {
    var handler = function () {
        requestAnimationFrame(callback);
    };
    elem.addEventListener('scroll', handler);
    return function () {
        elem.removeEventListener('scroll', handler);
    };
}
var Position = /** @class */ (function (_super) {
    __extends(Position, _super);
    function Position(props) {
        var _this = _super.call(this, props) || this;
        _this.getTarget = function () {
            var target = _this.props.target;
            var targetElement = typeof target === 'function' ? target() : target;
            return (targetElement && ReactDOM.findDOMNode(targetElement)) || null;
        };
        _this.maybeUpdatePosition = function (placementChanged) {
            var target = _this.getTarget();
            if (!_this.props.shouldUpdatePosition &&
                target === _this._lastTarget &&
                !placementChanged) {
                return;
            }
            _this.updatePosition(target);
        };
        _this.state = {
            ready: false,
            positionLeft: 0,
            positionTop: 0,
            arrowOffsetLeft: null,
            arrowOffsetTop: null
        };
        _this._lastTarget = null;
        _this.componentId = uuid();
        return _this;
    }
    Position.prototype.updatePosition = function (target) {
        var _this = this;
        var _a, _b;
        this._lastTarget = target;
        /** 标记宿主元素的PopOver祖先，用于后续判断PopOver 是否可以 root close */
        if (target) {
            var parentPopover = (_a = target === null || target === void 0 ? void 0 : target.closest) === null || _a === void 0 ? void 0 : _a.call(target, '[role=popover]');
            if (!this.parentPopover && parentPopover) {
                this.parentPopover = parentPopover;
                this.parentPopover.setAttribute(SubPopoverDisplayedID + '-' + this.componentId, true);
            }
        }
        if (!target || !target.offsetWidth) {
            // 靠这个 re-render 来重置 position
            return this.setState({});
        }
        var watchTargetSizeChange = this.props.watchTargetSizeChange;
        var overlay = this.overlay;
        var container = getContainer(this.props.container, ownerDocument(this).body);
        if (!this.watchedTarget || this.watchedTarget !== target) {
            (_b = this.resizeDispose) === null || _b === void 0 ? void 0 : _b.forEach(function (fn) { return fn(); });
            this.watchedTarget = target;
            this.resizeDispose = [
                watchTargetSizeChange !== false
                    ? resizeSensor(target, function () { return _this.updatePosition(target); })
                    : noop,
                resizeSensor(overlay, function () { return _this.updatePosition(target); })
            ];
            var scrollParent = getScrollParent(target);
            if (scrollParent && container.contains(scrollParent)) {
                this.resizeDispose.push(onScroll(scrollParent, function () {
                    _this.updatePosition(target);
                }));
            }
        }
        this.setState(__assign(__assign({}, calculatePosition(this.props.placement, overlay, target, container, this.props.containerPadding, this.props.offset)), { ready: true }));
    };
    Position.prototype.componentDidMount = function () {
        this.overlay = findDOMNode(this);
        this.updatePosition(this.getTarget());
    };
    Position.prototype.componentDidUpdate = function (prevProps) {
        this.maybeUpdatePosition(this.props.placement !== prevProps.placement);
    };
    Position.prototype.componentWillUnmount = function () {
        var _a;
        // 一个 PopOver 关闭时，需把挂载父 PopOver 的标记去掉
        // 这里可能会存在多个子 PopOver 的情况，所以需要加上 componentId
        if (this.parentPopover &&
            this.parentPopover.getAttribute(SubPopoverDisplayedID + '-' + this.componentId)) {
            this.parentPopover.removeAttribute(SubPopoverDisplayedID + '-' + this.componentId);
            this.parentPopover = null;
        }
        (_a = this.resizeDispose) === null || _a === void 0 ? void 0 : _a.forEach(function (fn) { return fn(); });
    };
    Position.prototype.render = function () {
        var _a = this.props, children = _a.children, className = _a.className, props = __rest(_a, ["children", "className"]);
        var _b = this.state, ready = _b.ready, positionLeft = _b.positionLeft, positionTop = _b.positionTop, arrowPosition = __rest(_b, ["ready", "positionLeft", "positionTop"]);
        // These should not be forwarded to the child.
        delete props.target;
        delete props.container;
        delete props.containerPadding;
        delete props.shouldUpdatePosition;
        var child = React.Children.only(children);
        return cloneElement(child, __assign(__assign(__assign(__assign({}, props), arrowPosition), (child.props.offset ? { offset: child.props.offset } : {})), { 
            // FIXME: Don't forward `positionLeft` and `positionTop` via both props
            // and `props.style`.
            positionLeft: positionLeft, positionTop: positionTop, className: cx(className, child.props.className), style: __assign(__assign({}, child.props.style), { left: positionLeft, top: positionTop, visibility: ready ? undefined : 'hidden' }), componentId: this.componentId }));
    };
    Position.defaultProps = {
        containerPadding: 0,
        placement: 'right',
        shouldUpdatePosition: false
    };
    return Position;
}(React.Component));
var Overlay = /** @class */ (function (_super) {
    __extends(Overlay, _super);
    function Overlay(props) {
        var _this = _super.call(this, props) || this;
        _this.position = null;
        _this.positionRef = function (position) {
            _this.position = position;
        };
        _this.state = {
            exited: !props.show
        };
        return _this;
    }
    Overlay.prototype.updatePosition = function () {
        var _a;
        (_a = this.position) === null || _a === void 0 ? void 0 : _a.maybeUpdatePosition(true);
    };
    Overlay.prototype.componentDidUpdate = function (prevProps, prevState) {
        var props = this.props;
        if (prevProps.show !== props.show && props.show) {
            this.setState({ exited: false });
        }
        else if (props.transition !== prevProps.transition && !props.transition) {
            // Otherwise let handleHidden take care of marking exited.
            this.setState({ exited: true });
        }
    };
    Overlay.prototype.onHiddenListener = function (node) {
        this.setState({ exited: true });
        if (this.props.onExited) {
            this.props.onExited(node);
        }
    };
    Overlay.prototype.getContainerSelector = function () {
        var containerSelector = this.props.containerSelector;
        var container = null;
        if (typeof containerSelector === 'string') {
            container = document.querySelector(containerSelector);
        }
        return container;
    };
    Overlay.prototype.render = function () {
        var _a;
        var _b = this.props, containerPadding = _b.containerPadding, target = _b.target, placement = _b.placement, shouldUpdatePosition = _b.shouldUpdatePosition, rootClose = _b.rootClose, children = _b.children; _b.watchTargetSizeChange; var Transition = _b.transition, offset = _b.offset, props = __rest(_b, ["containerPadding", "target", "placement", "shouldUpdatePosition", "rootClose", "children", "watchTargetSizeChange", "transition", "offset"]);
        var container = (this.getContainerSelector()
            ? this.getContainerSelector
            : this.props.container) || ((_a = this.context) === null || _a === void 0 ? void 0 : _a.getModalContainer);
        var mountOverlay = props.show || (Transition && !this.state.exited);
        if (!mountOverlay) {
            // Don't bother showing anything if we don't have to.
            return null;
        }
        var child = children;
        // Position is be inner-most because it adds inline styles into the child,
        // which the other wrappers don't forward correctly.
        child = (
        // @ts-ignore
        React.createElement(Position, __assign({}, {
            container: container,
            containerPadding: containerPadding,
            target: target,
            placement: placement,
            shouldUpdatePosition: shouldUpdatePosition,
            offset: offset
        }, { ref: this.positionRef }), child));
        if (Transition) {
            var onExit = props.onExit, onExiting = props.onExiting, onEnter = props.onEnter, onEntering = props.onEntering, onEntered = props.onEntered;
            // This animates the child node by injecting props, so it must precede
            // anything that adds a wrapping div.
            child = (React.createElement(Transition, { in: props.show, appear: true, onExit: onExit, onExiting: onExiting, onExited: this.onHiddenListener, onEnter: onEnter, onEntering: onEntering, onEntered: onEntered }, child));
        }
        // This goes after everything else because it adds a wrapping div.
        if (rootClose) {
            return (
            // @ts-ignore
            React.createElement(Portal, { container: container },
                React.createElement(RootClose, { onRootClose: props.onHide }, function (ref) {
                    if (React.isValidElement(child)) {
                        return React.cloneElement(child, {
                            ref: ref
                        });
                    }
                    return React.createElement("div", { ref: ref }, child);
                })));
        }
        // @ts-ignore
        return React.createElement(Portal, { container: container }, child);
    };
    Overlay.defaultProps = {
        placement: 'auto'
    };
    Overlay.contextType = EnvContext;
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [HTMLElement]),
        __metadata("design:returntype", void 0)
    ], Overlay.prototype, "onHiddenListener", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], Overlay.prototype, "getContainerSelector", null);
    return Overlay;
}(React.Component));

export { SubPopoverDisplayedID, Overlay as default };
