// office-viewer/lib/excel/render/autoFilter/sortState/SortButtonUI.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var H = require('../../../../util/H.js');

/**
 * 排序按钮
 */
var SortButtonUI = /** @class */ (function () {
    function SortButtonUI(container, text, onClick, active, iconSVG) {
        var _this = this;
        var element = H.H('div', {
            className: 'ov-excel-auto-filter__sort-button',
            parent: container
        });
        this.element = element;
        element.addEventListener('click', function () {
            onClick(_this.active);
        });
        var icon = H.H('div', {
            className: 'ov-excel-auto-filter__sort-button-icon',
            parent: element
        });
        icon.innerHTML = iconSVG;
        var textElement = H.H('div', {
            className: 'ov-excel-auto-filter__sort-button-text',
            parent: element
        });
        textElement.textContent = text;
        this.active = active;
    }
    SortButtonUI.prototype.setActive = function (active) {
        this.active = active;
        if (active) {
            this.element.classList.add('ov-excel-auto-filter__sort-button-active');
        }
        else {
            this.element.classList.remove('ov-excel-auto-filter__sort-button-active');
        }
    };
    return SortButtonUI;
}());

exports.SortButtonUI = SortButtonUI;
