// office-viewer/esm/excel/render/autoFilter/AutoFilterMenuUI.js
import { H } from '../../../util/H.js';
import { FilterColumnUI } from './filterColumn/FilterColumnUI.js';
import { SortStateUI } from './sortState/SortStateUI.js';

var AutoFilterMenuUI = /** @class */ (function () {
    function AutoFilterMenuUI(autoFilterIcon, container, sheet, autoFilter, rangeRef, colIndex, headerRowCount) {
        if (headerRowCount === void 0) { headerRowCount = 1; }
        var menuContainer = H('div', {
            className: 'ov-excel-auto-filter-menu',
            parent: container
        });
        this.menuContainer = menuContainer;
        new SortStateUI(autoFilterIcon, menuContainer, sheet, autoFilter, rangeRef, colIndex, headerRowCount);
        new FilterColumnUI(autoFilterIcon, menuContainer, sheet, autoFilter, rangeRef, colIndex, headerRowCount);
        this.autoFilter = autoFilter;
        this.colIndex = colIndex;
        this.sheet = sheet;
        menuContainer.addEventListener('mousedown', function (e) {
            e.stopPropagation();
        });
        menuContainer.addEventListener('wheel', function (e) {
            e.stopPropagation();
        });
    }
    AutoFilterMenuUI.prototype.hide = function () {
        this.menuContainer.style.display = 'none';
    };
    AutoFilterMenuUI.prototype.show = function () {
        this.menuContainer.style.display = 'flex';
    };
    return AutoFilterMenuUI;
}());

export { AutoFilterMenuUI };
