// office-viewer/lib/UnSupport.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');

/**
 * 不支持的文件类型，主要用于显示报错信息
 */
var UnSupport = /** @class */ (function () {
    function UnSupport(errorMessage) {
        this.errorMessage = errorMessage;
    }
    UnSupport.prototype.updateOptions = function (options) {
        throw new Error('Method not implemented.');
    };
    UnSupport.prototype.updateVariable = function () {
        throw new Error('Method not implemented.');
    };
    UnSupport.prototype.render = function (root, options) {
        return tslib.__awaiter(this, void 0, void 0, function () {
            return tslib.__generator(this, function (_a) {
                root.innerHTML = this.errorMessage;
                return [2 /*return*/];
            });
        });
    };
    UnSupport.prototype.download = function (fileName) {
        throw new Error('Method not implemented.');
    };
    UnSupport.prototype.print = function () {
        throw new Error('Method not implemented.');
    };
    UnSupport.prototype.destroy = function () {
        throw new Error('Method not implemented.');
    };
    return UnSupport;
}());

exports["default"] = UnSupport;
