// amis-core/lib/components/LazyComponent.js
/**
 * amis-core v6.13.0
 * Copyright 2018-2025 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var React = require('react');
var reactIntersectionObserver = require('react-intersection-observer');
var theme = require('../theme.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var __react_jsx__ = require('react');
var _J$X_ = (__react_jsx__["default"] || __react_jsx__).createElement;
(__react_jsx__["default"] || __react_jsx__).Fragment;
var LazyComponent = /** @class */ (function (_super) {
    tslib.__extends(LazyComponent, _super);
    function LazyComponent(props) {
        var _this = this;
        var _a;
        _this = _super.call(this, props) || this;
        _this.mounted = false;
        _this.handleVisibleChange = _this.handleVisibleChange.bind(_this);
        _this.mounted = true;
        _this.state = {
            visible: (_a = props.defaultVisible) !== null && _a !== void 0 ? _a : false,
            component: props.component
        };
        return _this;
    }
    LazyComponent.prototype.componentDidMount = function () {
        // jest 里面有点异常，先手动让它总是可见
        if (typeof jest !== 'undefined' || this.state.visible) {
            this.handleVisibleChange(true);
        }
    };
    LazyComponent.prototype.componentWillUnmount = function () {
        this.mounted = false;
    };
    LazyComponent.prototype.handleVisibleChange = function (visible, entry) {
        var _this = this;
        this.setState({
            visible: visible
        });
        if (!visible || this.state.component || !this.props.getComponent) {
            return;
        }
        this.props
            .getComponent()
            .then(function (component) {
            return _this.mounted &&
                typeof component === 'function' &&
                _this.setState({
                    component: component
                });
        })
            .catch(function (reason) {
            return _this.mounted &&
                _this.setState({
                    component: function () { return (_J$X_("div", { className: "alert alert-danger" }, String(reason))); }
                });
        });
    };
    LazyComponent.prototype.render = function () {
        var _a = this.props, placeholder = _a.placeholder, unMountOnHidden = _a.unMountOnHidden, childProps = _a.childProps, partialVisibility = _a.partialVisibility, children = _a.children, className = _a.className, rest = tslib.__rest(_a, ["placeholder", "unMountOnHidden", "childProps", "partialVisibility", "children", "className"]);
        var cx = this.props.classnames;
        var _b = this.state, visible = _b.visible, Component = _b.component;
        // 需要监听从可见到不可见。
        if (unMountOnHidden) {
            return (_J$X_(reactIntersectionObserver.InView, { onChange: this.handleVisibleChange, threshold: partialVisibility ? 0 : 1 }, function (_a) {
                var ref = _a.ref;
                return (_J$X_("div", { ref: ref, className: "visibility-sensor ".concat(visible ? 'in' : '') }, Component && visible ? (_J$X_(Component, tslib.__assign({}, rest, childProps))) : children && visible ? (children) : (placeholder)));
            }));
        }
        if (!visible) {
            return (_J$X_(reactIntersectionObserver.InView, { onChange: this.handleVisibleChange, threshold: partialVisibility ? 0 : 1 }, function (_a) {
                var ref = _a.ref;
                return (_J$X_("div", { ref: ref, className: "visibility-sensor" }, placeholder));
            }));
        }
        else if (Component) {
            // 只监听不可见到可见，一旦可见了，就销毁检查。
            return _J$X_(Component, tslib.__assign({}, rest, childProps));
        }
        else if (children) {
            return children;
        }
        return _J$X_("div", { className: cx('LazyComponent', className) }, placeholder);
    };
    LazyComponent.defaultProps = {
        placeholder: _J$X_("span", null, "Loading..."),
        unMountOnHidden: false,
        partialVisibility: true
    };
    return LazyComponent;
}(React__default["default"].Component));
var themedLazyComponent = theme.themeable(LazyComponent);
themedLazyComponent.defaultProps = LazyComponent.defaultProps;

exports.LazyComponent = LazyComponent;
exports["default"] = themedLazyComponent;
