// amis-core/lib/actions/Action.js
/**
 * amis-core v6.13.0
 * Copyright 2018-2025 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var omit = require('lodash/omit');
var tpl = require('../utils/tpl.js');
require('amis-formula');
require('moment');
var object = require('../utils/object.js');
require('../utils/memoryParse.js');
require('@rc-component/mini-decimal');
var dataMapping = require('../utils/dataMapping.js');
require('../utils/filter.js');
var debug = require('../utils/debug.js');
require('../utils/api.js');
require('file-saver');
require('../utils/browser.js');
require('../utils/ColorScale.js');
require('../utils/columnsSplit.js');
require('../utils/DataSchema.js');
require('../utils/DataScope.js');
require('../utils/dom.js');
require('../utils/errors.js');
require('../utils/helper.js');
require('lodash/isPlainObject');
require('lodash/isObject');
require('lodash/isString');
require('lodash/isBoolean');
require('../utils/highlight.js');
require('../utils/icon.js');
require('../utils/image.js');
require('lodash/isEqual');
require('lodash/debounce');
require('../utils/resize-sensor.js');
require('react');
require('react-overlays/useRootClose');
require('react-dom');
require('../utils/SimpleMap.js');
require('lodash/mapValues');
require('lodash/camelCase');
require('lodash/cloneDeep');
require('lodash/map');
require('lodash/isEmpty');
require('lodash/kebabCase');
require('uncontrollable');
require('hoist-non-react-statics');
require('../utils/validations.js');
require('../utils/resolveCondition.js');
require('mobx');
require('../utils/Animation.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var omit__default = /*#__PURE__*/_interopDefaultLegacy(omit);

// 循环动作执行状态
exports.LoopStatus = void 0;
(function (LoopStatus) {
    LoopStatus[LoopStatus["NORMAL"] = 0] = "NORMAL";
    LoopStatus[LoopStatus["BREAK"] = 1] = "BREAK";
    LoopStatus[LoopStatus["CONTINUE"] = 2] = "CONTINUE";
})(exports.LoopStatus || (exports.LoopStatus = {}));
// 存储 Action 和类型的映射关系，用于后续查找
var ActionTypeMap = {};
// 存储动作属性排除列表
var ActionIgnoreKey = {};
// 存储组件专有动作的属性排除列表
var CmptIgnoreMap = {};
// 注册 Action
var registerAction = function (type, action) {
    ActionTypeMap[type] = action;
};
// 通过类型获取 Action 实例
var getActionByType = function (type) {
    return ActionTypeMap[type];
};
// 根据动作类型获取属性排除列表
var getOmitActionProp = function (type) {
    var omitList = [];
    switch (type) {
        case 'toast':
            omitList = [
                'msgType',
                'msg',
                'position',
                'closeButton',
                'showIcon',
                'timeout',
                'title'
            ];
            break;
        case 'alert':
            omitList = ['msg'];
            break;
        case 'confirm':
            omitList = ['msg', 'title'];
            break;
        case 'ajax':
        case 'download':
            omitList = ['api', 'messages', 'options'];
            break;
        case 'setValue':
            omitList = ['value', 'index'];
            break;
        case 'copy':
            omitList = ['content', 'copyFormat'];
            break;
        case 'email':
            omitList = ['to', 'cc', 'bcc', 'subject', 'body'];
            break;
        case 'link':
            omitList = ['link', 'blank', 'params'];
            break;
        case 'url':
            omitList = ['url', 'blank', 'params'];
            break;
        case 'for':
            omitList = ['loopName'];
            break;
        case 'goPage':
            omitList = ['delta'];
            break;
        case 'custom':
            omitList = ['script'];
            break;
        case 'broadcast':
            omitList = ['eventName'];
            break;
        case 'dialog':
            omitList = ['dialog'];
            break;
        case 'drawer':
            omitList = ['drawer'];
            break;
        case 'confirmDialog':
            omitList = ['dialog'];
            break;
        case 'reload':
            omitList = ['resetPage'];
            break;
    }
    return omitList;
};
var getTargetComponent = function (action, renderer, event, key) {
    var targetComponent = renderer;
    if (key && event.context.scoped) {
        var func = action.componentId ? 'getComponentById' : 'getComponentByName';
        if (typeof event.context.scoped[func] === 'function') {
            targetComponent = event.context.scoped[func](key);
        }
    }
    return targetComponent;
};
var runActions = function (actions, renderer, event) { return tslib.__awaiter(void 0, void 0, void 0, function () {
    var actions_1, actions_1_1, actionConfig, actionInstrance, e_1, ignore, e_2_1;
    var e_2, _a;
    var _b;
    return tslib.__generator(this, function (_c) {
        switch (_c.label) {
            case 0:
                if (!Array.isArray(actions)) {
                    actions = [actions];
                }
                _c.label = 1;
            case 1:
                _c.trys.push([1, 9, 10, 11]);
                actions_1 = tslib.__values(actions), actions_1_1 = actions_1.next();
                _c.label = 2;
            case 2:
                if (!!actions_1_1.done) return [3 /*break*/, 8];
                actionConfig = actions_1_1.value;
                actionInstrance = getActionByType(actionConfig.actionType);
                // 如果存在指定组件ID，说明是组件专有动作
                if (!actionInstrance &&
                    (actionConfig.componentId || actionConfig.componentName)) {
                    actionInstrance = [
                        'static',
                        'nonstatic',
                        'show',
                        'visibility',
                        'hidden',
                        'enabled',
                        'disabled',
                        'usability'
                    ].includes(actionConfig.actionType)
                        ? getActionByType('status')
                        : getActionByType('component');
                }
                else if (['url', 'link', 'jump'].includes(actionConfig.actionType)) {
                    // 打开页面动作
                    actionInstrance = getActionByType('openlink');
                }
                // 找不到就通过组件专有动作完成
                if (!actionInstrance) {
                    actionInstrance = getActionByType('component');
                }
                _c.label = 3;
            case 3:
                _c.trys.push([3, 5, , 6]);
                // 这些节点的子节点运行逻辑由节点内部实现
                return [4 /*yield*/, runAction(actionInstrance, actionConfig, renderer, event)];
            case 4:
                // 这些节点的子节点运行逻辑由节点内部实现
                _c.sent();
                return [3 /*break*/, 6];
            case 5:
                e_1 = _c.sent();
                ignore = (_b = actionConfig.ignoreError) !== null && _b !== void 0 ? _b : false;
                if (!ignore) {
                    // 通过标记 stop 来阻止后续动作执行
                    // 不要抛出，避免后续的代码逻辑不执行，同时避免 unhandled promise rejection
                    event.stopPropagation();
                    event.preventDefault();
                    console.error("".concat(actionConfig.actionType, " \u52A8\u4F5C\u6267\u884C\u5931\u8D25\uFF0C\u539F\u56E0\uFF1A").concat(e_1.message || '未知'));
                    // throw Error(
                    //   `${actionConfig.actionType} 动作执行失败，原因：${
                    //     e.message || '未知'
                    //   }`
                    // );
                }
                return [3 /*break*/, 6];
            case 6:
                if (event.stoped) {
                    return [3 /*break*/, 8];
                }
                _c.label = 7;
            case 7:
                actions_1_1 = actions_1.next();
                return [3 /*break*/, 2];
            case 8: return [3 /*break*/, 11];
            case 9:
                e_2_1 = _c.sent();
                e_2 = { error: e_2_1 };
                return [3 /*break*/, 11];
            case 10:
                try {
                    if (actions_1_1 && !actions_1_1.done && (_a = actions_1.return)) _a.call(actions_1);
                }
                finally { if (e_2) throw e_2.error; }
                return [7 /*endfinally*/];
            case 11: return [2 /*return*/];
        }
    });
}); };
// 执行动作，与原有动作处理打通
var runAction = function (actionInstrance, actionConfig, renderer, event) { return tslib.__awaiter(void 0, void 0, void 0, function () {
    var additional, action, rendererProto, mergeData, expression, isStop, preventDefault, key, api, cmptFlag, targetComponent, args, afterMappingData, actionData, data, stopped, actionResult, stopPropagation;
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x;
    return tslib.__generator(this, function (_y) {
        switch (_y.label) {
            case 0:
                additional = {
                    event: event
                };
                action = tslib.__assign({}, actionConfig);
                action.args = tslib.__assign({}, actionConfig.args);
                rendererProto = (_c = (_b = (_a = renderer.props).getData) === null || _b === void 0 ? void 0 : _b.call(_a)) !== null && _c !== void 0 ? _c : renderer.props.data;
                // __rendererData默认为renderer.props.data，兼容表单项值变化时的data读取
                if (!((_d = event.data) === null || _d === void 0 ? void 0 : _d.__rendererData)) {
                    additional = {
                        event: event,
                        __rendererData: rendererProto // 部分组件交互后会有更新，如果想要获取那部分数据，可以通过事件数据获取
                    };
                }
                mergeData = object.injectObjectChain(event.data, additional);
                expression = (_e = action.expression) !== null && _e !== void 0 ? _e : action.execOn;
                isStop = false;
                if (!expression) return [3 /*break*/, 2];
                return [4 /*yield*/, tpl.evalExpressionWithConditionBuilderAsync(expression, mergeData, true)];
            case 1:
                isStop = !(_y.sent());
                _y.label = 2;
            case 2:
                if (isStop) {
                    return [2 /*return*/];
                }
                preventDefault = false;
                if (!action.preventDefault) return [3 /*break*/, 4];
                return [4 /*yield*/, tpl.evalExpressionWithConditionBuilderAsync(action.preventDefault, mergeData, false)];
            case 3:
                preventDefault = _y.sent();
                _y.label = 4;
            case 4:
                key = {
                    componentId: dataMapping.dataMapping(action.componentId, mergeData),
                    componentName: dataMapping.dataMapping(action.componentName, mergeData)
                };
                // 兼容args包裹的用法
                if (action.actionType === 'dialog') {
                    action.dialog = tslib.__assign({}, ((_f = action.dialog) !== null && _f !== void 0 ? _f : (_g = action.args) === null || _g === void 0 ? void 0 : _g.dialog));
                    (_h = action.args) === null || _h === void 0 ? true : delete _h.dialog;
                }
                else if (action.actionType === 'drawer') {
                    action.drawer = tslib.__assign({}, ((_j = action.drawer) !== null && _j !== void 0 ? _j : (_k = action.args) === null || _k === void 0 ? void 0 : _k.drawer));
                    (_l = action.args) === null || _l === void 0 ? true : delete _l.drawer;
                }
                else if (['ajax', 'download'].includes(action.actionType)) {
                    api = (_m = action.api) !== null && _m !== void 0 ? _m : (_o = action.args) === null || _o === void 0 ? void 0 : _o.api;
                    action.api = typeof api === 'string' ? api : tslib.__assign({}, api);
                    action.options = tslib.__assign({}, ((_p = action.options) !== null && _p !== void 0 ? _p : (_q = action.args) === null || _q === void 0 ? void 0 : _q.options));
                    action.messages = tslib.__assign({}, ((_r = action.messages) !== null && _r !== void 0 ? _r : (_s = action.args) === null || _s === void 0 ? void 0 : _s.messages));
                    (_t = action.args) === null || _t === void 0 ? true : delete _t.api;
                    (_u = action.args) === null || _u === void 0 ? true : delete _u.options;
                    (_v = action.args) === null || _v === void 0 ? true : delete _v.messages;
                }
                cmptFlag = key.componentId || key.componentName;
                targetComponent = getTargetComponent(action, renderer, event, cmptFlag);
                args = dataMapping.dataMapping(action.args, mergeData, function (key) {
                    var _a;
                    var curCmptType = (_a = targetComponent === null || targetComponent === void 0 ? void 0 : targetComponent.props) === null || _a === void 0 ? void 0 : _a.type;
                    var curActionType = action.actionType;
                    var ignoreKey = tslib.__spreadArray(tslib.__spreadArray([], tslib.__read((ActionIgnoreKey[curActionType] || [])), false), tslib.__read((CmptIgnoreMap[curCmptType] || [])), false);
                    return ignoreKey.includes(key);
                });
                afterMappingData = dataMapping.dataMapping(action.data, mergeData);
                actionData = args && Object.keys(args).length
                    ? omit__default["default"](tslib.__assign(tslib.__assign({}, args), (afterMappingData !== null && afterMappingData !== void 0 ? afterMappingData : {})), getOmitActionProp(action.actionType))
                    : afterMappingData;
                data = actionData !== undefined &&
                    !['ajax', 'download', 'dialog', 'drawer'].includes(action.actionType) // 避免非法配置影响对actionData的判断，导致动作配置中的数据映射失败
                    ? actionData
                    : mergeData;
                (_w = console.group) === null || _w === void 0 ? void 0 : _w.call(console, "run action ".concat(action.actionType));
                console.debug("[".concat(action.actionType, "] action args, data"), args, data);
                debug.debug('action', "run action ".concat(action.actionType, " with args"), args);
                debug.debug('action', "run action ".concat(action.actionType, " with data"), data);
                _y.label = 5;
            case 5:
                _y.trys.push([5, , 9, 10]);
                stopped = false;
                return [4 /*yield*/, actionInstrance.run(tslib.__assign(tslib.__assign(tslib.__assign({}, action), { args: args, rawData: actionConfig.data, data: action.actionType === 'reload' ? actionData : data }), key), renderer, event, mergeData)];
            case 6:
                actionResult = _y.sent();
                // 二次确认弹窗如果取消，则终止后续动作
                if ((action === null || action === void 0 ? void 0 : action.actionType) === 'confirmDialog' && !actionResult) {
                    stopped = true;
                    preventDefault = true; // 这种对表单项change比较有意义，例如switch切换时弹确认弹窗，如果取消后不能把switch修改了
                }
                stopPropagation = false;
                if (!action.stopPropagation) return [3 /*break*/, 8];
                return [4 /*yield*/, tpl.evalExpressionWithConditionBuilderAsync(action.stopPropagation, mergeData, false)];
            case 7:
                stopPropagation = _y.sent();
                _y.label = 8;
            case 8:
                // 阻止原有动作执行
                preventDefault && event.preventDefault();
                // 阻止后续动作执行
                (stopPropagation || stopped) && event.stopPropagation();
                return [3 /*break*/, 10];
            case 9:
                console.debug("[".concat(action.actionType, "] action end event"), event);
                (_x = console.groupEnd) === null || _x === void 0 ? void 0 : _x.call(console);
                return [7 /*endfinally*/];
            case 10: return [2 /*return*/];
        }
    });
}); };
// 注册动作参数映射忽略键
var registerActionMappingIgnoreKey = function (actionType, ignoreKey, replace) {
    if (replace === void 0) { replace = true; }
    if (replace) {
        ActionIgnoreKey[actionType] = ignoreKey;
        return;
    }
    ActionIgnoreKey[actionType] = tslib.__spreadArray(tslib.__spreadArray([], tslib.__read((ActionIgnoreKey[actionType] || [])), false), tslib.__read(ignoreKey), false);
};
// 注册多个动作参数映射忽略键
var registerActionMappingIgnoreMap = function (maps, replace) {
    if (replace === void 0) { replace = true; }
    Object.keys(maps).forEach(function (key) {
        registerActionMappingIgnoreKey(key, maps[key], replace);
    });
};
// 注册组件动作参数映射忽略键
var registerComponentActionMappingIgnoreKey = function (cmptType, ignoreKey, replace) {
    if (replace === void 0) { replace = true; }
    if (replace) {
        CmptIgnoreMap[cmptType] = ignoreKey;
        return;
    }
    CmptIgnoreMap[cmptType] = tslib.__spreadArray(tslib.__spreadArray([], tslib.__read((CmptIgnoreMap[cmptType] || [])), false), tslib.__read(ignoreKey), false);
};
// 注册多个组件动作参数映射忽略键
var registerComponentActionMappingIgnoreMap = function (maps, replace) {
    if (replace === void 0) { replace = true; }
    Object.keys(maps).forEach(function (key) {
        registerComponentActionMappingIgnoreKey(key, maps[key], replace);
    });
};
// 注册默认忽略键（将来是否需要移到相应的模块中）
registerActionMappingIgnoreMap({
    ajax: ['adaptor', 'responseAdaptor', 'requestAdaptor', 'responseData']
});
// 注册组件默认忽略键（将来是否需要移到相应的渲染器中）
registerComponentActionMappingIgnoreMap({
    'input-table': ['condition'],
    'table': ['condition'],
    'table2': ['condition'],
    'crud': ['condition'],
    'combo': ['condition'],
    'list': ['condition'],
    'cards': ['condition']
});

exports.getActionByType = getActionByType;
exports.getTargetComponent = getTargetComponent;
exports.registerAction = registerAction;
exports.registerActionMappingIgnoreKey = registerActionMappingIgnoreKey;
exports.registerActionMappingIgnoreMap = registerActionMappingIgnoreMap;
exports.registerComponentActionMappingIgnoreKey = registerComponentActionMappingIgnoreKey;
exports.registerComponentActionMappingIgnoreMap = registerComponentActionMappingIgnoreMap;
exports.runAction = runAction;
exports.runActions = runActions;
