// office-viewer/esm/excel/render/cell/frozen/getFrozenLeftViewPointRange.js
import { getRange } from '../../../sheet/getViewRange.js';
import { getFrozenRange } from './getFrozenRange.js';

/**
 * 获取左边冻结的范围
 * @param xSplit
 * @param height
 * @param scrollTop
 * @param getRowHeight
 * @param getColWidth
 * @param rowPositionCache
 * @param rowHiddenRange
 * @returns
 */
function getFrozenLeftViewPointRange(xSplit, height, scrollTop, leftShift, topShift, getRowHeight, getColWidth, rowPositionCache) {
    if (rowPositionCache === void 0) { rowPositionCache = []; }
    var _a = getRange(scrollTop, topShift, height, getRowHeight, [], rowPositionCache), rows = _a.indexes, startRowOffset = _a.startOffset, rowSizes = _a.sizes;
    var _b = getFrozenRange(xSplit, leftShift, getColWidth), cols = _b.indexes, startColOffset = _b.startOffset, colSizes = _b.sizes, width = _b.length;
    return {
        region: 'left-frozen',
        rows: rows,
        rowSizes: rowSizes,
        height: height,
        startRowOffset: startRowOffset,
        cols: cols,
        colSizes: colSizes,
        width: width,
        startColOffset: startColOffset
    };
}

export { getFrozenLeftViewPointRange };
