// office-viewer/lib/excel/render/dnd/handleMousedown.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var handleMousemove = require('./handleMousemove.js');
var handleMouseup = require('./handleMouseup.js');
var mousedownColHeader = require('./mousedownColHeader.js');
var mousedownCorner = require('./mousedownCorner.js');
var mousedownRowHeader = require('./mousedownRowHeader.js');
var mousedownCell = require('./mousedownCell.js');

/**
 * 处理拖拽事件，包括单元格和表头部分
 */
var DRAG_STATE_DEFAULT = {
    isDragging: false,
    dragStart: {
        pageX: 0,
        pageY: 0,
        offsetX: 0,
        offsetY: 0
    },
    row: 0,
    col: 0,
    tmpRowHeight: 0,
    tmpColWidth: 0,
    region: 'normal',
    dragType: 'cell'
};
var dragState = JSON.parse(JSON.stringify(DRAG_STATE_DEFAULT));
function resetDragState() {
    Object.assign(dragState, DRAG_STATE_DEFAULT);
}
/**
 * 鼠标按下事件
 */
function handleMousedown(workbook, hitTestResult, container, mouseEvent) {
    // 只处理左键
    if (mouseEvent.button !== 0) {
        return;
    }
    var newSelection = null;
    if (hitTestResult.type === 'cell') {
        newSelection = mousedownCell.mousedownCell(workbook, hitTestResult);
    }
    if (hitTestResult.type === 'row-header') {
        newSelection = mousedownRowHeader.mousedownRowHeader(workbook, hitTestResult);
    }
    if (hitTestResult.type === 'col-header') {
        newSelection = mousedownColHeader.mousedownColHeader(workbook, hitTestResult);
    }
    if (hitTestResult.type === 'corner') {
        newSelection = mousedownCorner.mousedownCorner(workbook);
    }
    var zoomLevel = workbook.getActiveSheet().getZoomLevel();
    dragState.workbook = workbook;
    dragState.container = container;
    dragState.region = hitTestResult.region;
    dragState.dragStart = {
        pageX: mouseEvent.pageX,
        pageY: mouseEvent.pageY,
        offsetX: mouseEvent.offsetX / zoomLevel,
        offsetY: mouseEvent.offsetY / zoomLevel
    };
    dragState.row = hitTestResult.startRow;
    dragState.col = hitTestResult.startCol;
    dragState.dragType = hitTestResult.type;
    if (newSelection) {
        dragState.selection = newSelection;
        workbook.uiEvent.emit('CHANGE_SELECTION', newSelection);
    }
    document.addEventListener('mousemove', handleMousemove.handleMousemove, true);
    document.addEventListener('touchmove', handleMousemove.handleMousemove, true);
    document.addEventListener('mouseup', handleMouseup.handleMouseup);
    document.addEventListener('touchend', handleMouseup.handleMouseup);
}

exports.dragState = dragState;
exports.handleMousedown = handleMousedown;
exports.resetDragState = resetDragState;
