// amis-core/lib/utils/tpl-builtin.js
/**
 * amis-core v6.13.0
 * Copyright 2018-2025 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var amisFormula = require('amis-formula');
require('moment');
require('lodash/isPlainObject');
require('./memoryParse.js');
var resolveVariableAndFilter = require('./resolveVariableAndFilter.js');
require('@rc-component/mini-decimal');
var tokenize = require('./tokenize.js');
var resolveVariableAndFilterForAsync = require('./resolveVariableAndFilterForAsync.js');
require('./filter.js');

function matchSynatax(str) {
    var from = 0;
    while (true) {
        var idx = str.indexOf('$', from);
        if (~idx) {
            var nextToken = str[idx + 1];
            // 如果没有下一个字符，或者下一个字符是引号或者空格
            // 这个一般不是取值用法
            if (!nextToken || ~['"', "'", ' '].indexOf(nextToken)) {
                from = idx + 1;
                continue;
            }
            // 如果上个字符是转义也不是取值用法
            var prevToken = str[idx - 1];
            if (prevToken && prevToken === '\\') {
                from = idx + 1;
                continue;
            }
            return true;
        }
        else {
            break;
        }
    }
    return false;
}
function register() {
    var _this = this;
    return {
        name: 'builtin',
        test: function (str) { return typeof str === 'string' && matchSynatax(str); },
        removeEscapeToken: function (str) {
            return typeof str === 'string' ? str.replace(/\\\$/g, '$') : str;
        },
        compile: function (str, data, defaultFilter) {
            if (defaultFilter === void 0) { defaultFilter = '| html'; }
            try {
                return tokenize.tokenize(str, data, defaultFilter);
            }
            catch (e) {
                return "error: ".concat(e.message);
            }
        },
        asyncCompile: function (str, data, defaultFilter) {
            if (defaultFilter === void 0) { defaultFilter = '| html'; }
            return tslib.__awaiter(_this, void 0, void 0, function () {
                return tslib.__generator(this, function (_a) {
                    try {
                        return [2 /*return*/, resolveVariableAndFilterForAsync.resolveVariableAndFilterForAsync(str, data, defaultFilter)];
                    }
                    catch (e) {
                        return [2 /*return*/, "error: ".concat(e.message)];
                    }
                    return [2 /*return*/];
                });
            });
        }
    };
}
// 避免 resolveVariableAndFilter 被摇了
var testResolveVariableAndFilter = resolveVariableAndFilter.resolveVariableAndFilter;

Object.defineProperty(exports, 'getFilters', {
    enumerable: true,
    get: function () { return amisFormula.getFilters; }
});
Object.defineProperty(exports, 'registerFilter', {
    enumerable: true,
    get: function () { return amisFormula.registerFilter; }
});
Object.defineProperty(exports, 'registerFunction', {
    enumerable: true,
    get: function () { return amisFormula.registerFunction; }
});
exports.resolveVariableAndFilter = resolveVariableAndFilter.resolveVariableAndFilter;
exports.tokenize = tokenize.tokenize;
exports.resolveVariableAndFilterForAsync = resolveVariableAndFilterForAsync.resolveVariableAndFilterForAsync;
exports.register = register;
exports.testResolveVariableAndFilter = testResolveVariableAndFilter;
