// amis-core/lib/utils/tpl.js
/**
 * amis-core v6.13.0
 * Copyright 2018-2025 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var tplBuiltin = require('./tpl-builtin.js');
var tplLodash = require('./tpl-lodash.js');
var amisFormula = require('amis-formula');
var resolveCondition = require('./resolveCondition.js');
var tokenize = require('./tokenize.js');

var enginers = {};
function registerTplEnginer(name, enginer) {
    enginers[name] = enginer;
}
function filter(tpl, data) {
    if (data === void 0) { data = {}; }
    var rest = [];
    for (var _i = 2; _i < arguments.length; _i++) {
        rest[_i - 2] = arguments[_i];
    }
    if (!tpl || typeof tpl !== 'string') {
        return '';
    }
    var keys = Object.keys(enginers);
    for (var i = 0, len = keys.length; i < len; i++) {
        var enginer = enginers[keys[i]];
        if (enginer.test(tpl)) {
            return enginer.compile.apply(enginer, tslib.__spreadArray([tpl, data], tslib.__read(rest), false));
        }
        else if (enginer.removeEscapeToken) {
            tpl = enginer.removeEscapeToken(tpl);
        }
    }
    return tpl;
}
function asyncFilter(tpl, data) {
    if (data === void 0) { data = {}; }
    var rest = [];
    for (var _i = 2; _i < arguments.length; _i++) {
        rest[_i - 2] = arguments[_i];
    }
    if (!tpl || typeof tpl !== 'string') {
        return Promise.resolve('');
    }
    var keys = Object.keys(enginers);
    for (var i = 0, len = keys.length; i < len; i++) {
        var enginer = enginers[keys[i]];
        if (enginer.test(tpl)) {
            return enginer.asyncCompile.apply(enginer, tslib.__spreadArray([tpl, data], tslib.__read(rest), false));
        }
        else if (enginer.removeEscapeToken) {
            tpl = enginer.removeEscapeToken(tpl);
        }
    }
    return tpl;
}
// 缓存一下提升性能
var EVAL_CACHE = {};
var WARNING_CACHE = {};
var customEvalExpressionFn;
function setCustomEvalExpression(fn) {
    customEvalExpressionFn = fn;
}
// 几乎所有的 visibleOn requiredOn 都是通过这个方法判断出来结果，很粗暴也存在风险，建议自己实现。
// 如果想自己实现，请通过 setCustomEvalExpression 来替换。
function evalExpression(expression, data) {
    if (typeof customEvalExpressionFn === 'function') {
        return customEvalExpressionFn(expression, data);
    }
    if (!expression || typeof expression !== 'string') {
        return false;
    }
    /* jshint evil:true */
    try {
        if (typeof expression === 'string' &&
            expression.substring(0, 2) === '${' &&
            expression[expression.length - 1] === '}') {
            // 启用新版本的公式表达式
            return !!evalFormula(expression, data);
        }
        // 后续改用 FormulaExec['js']
        var debug_1 = false;
        var idx = expression.indexOf('debugger');
        if (~idx) {
            debug_1 = true;
            expression = expression.replace(/debugger;?/, '');
        }
        var fn = void 0;
        if (expression in EVAL_CACHE) {
            fn = EVAL_CACHE[expression];
        }
        else {
            fn = new Function('data', 'utils', "with(data) {".concat(debug_1 ? 'debugger;' : '', "return !!(").concat(expression, ");}"));
            EVAL_CACHE[expression] = fn;
        }
        data = data || {};
        return fn.call(data, data, amisFormula.getFilters());
    }
    catch (e) {
        if (!WARNING_CACHE[expression]) {
            console.warn(expression, e);
            WARNING_CACHE[expression] = true;
        }
        return false;
    }
}
/**
 * 解析表达式（支持condition-builder）
 * @param expression 表达式 or condition-builder对象
 * @param data 上下文
 * @returns
 */
function evalExpressionWithConditionBuilderAsync(expression, data, defaultResult) {
    return tslib.__awaiter(this, void 0, void 0, function () {
        return tslib.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!(Object.prototype.toString.call(expression) === '[object Object]')) return [3 /*break*/, 2];
                    return [4 /*yield*/, resolveCondition.resolveConditionAsync(expression, data, defaultResult)];
                case 1: return [2 /*return*/, _a.sent()];
                case 2: return [2 /*return*/, evalExpression(String(expression), data)];
            }
        });
    });
}
/**
 * 解析表达式（支持condition-builder）
 * @param expression 表达式 or condition-builder对象
 * @param data 上下文
 * @returns
 */
function evalExpressionWithConditionBuilder(expression, data, defaultResult) {
    // 支持ConditionBuilder
    if (Object.prototype.toString.call(expression) === '[object Object]') {
        return resolveCondition.resolveCondition(expression, data, defaultResult);
    }
    return evalExpression(String(expression), data);
}
function evalFormula(expression, data) {
    var ast = tokenize.memoParse(expression, {
        evalMode: false
    });
    return amisFormula.evaluate(ast, data, {
        defaultFilter: 'raw'
    });
}
var customEvalJsFn;
function setCustomEvalJs(fn) {
    customEvalJsFn = fn;
}
// 这个主要用在 formula 里面，用来动态的改变某个值。也很粗暴，建议自己实现。
// 如果想自己实现，请通过 setCustomEvalJs 来替换。
function evalJS(js, data) {
    if (typeof customEvalJsFn === 'function') {
        return customEvalJsFn(js, data);
    }
    /* jshint evil:true */
    try {
        if (typeof js === 'string' &&
            js.substring(0, 2) === '${' &&
            js[js.length - 1] === '}') {
            // 启用新版本的公式表达式
            return evalFormula(js, data);
        }
        var fn = new Function('data', 'utils', "with(data) {".concat(/^\s*return\b/.test(js) ? '' : 'return ').concat(js, ";}"));
        data = data || {};
        return fn.call(data, data, amisFormula.getFilters());
    }
    catch (e) {
        console.warn(js, e);
        return null;
    }
}
[tplBuiltin.register, tplLodash.register].forEach(function (fn) {
    if (!fn)
        return;
    var info = fn();
    registerTplEnginer(info.name, {
        test: info.test,
        compile: info.compile,
        asyncCompile: info.asyncCompile,
        removeEscapeToken: info.removeEscapeToken
    });
});

exports.asyncFilter = asyncFilter;
exports.evalExpression = evalExpression;
exports.evalExpressionWithConditionBuilder = evalExpressionWithConditionBuilder;
exports.evalExpressionWithConditionBuilderAsync = evalExpressionWithConditionBuilderAsync;
exports.evalJS = evalJS;
exports.filter = filter;
exports.registerTplEnginer = registerTplEnginer;
exports.setCustomEvalExpression = setCustomEvalExpression;
exports.setCustomEvalJs = setCustomEvalJs;
